// @douyinfe/semi-foundation/lib/es/checkbox/checkboxFoundation.js
import _Object$assign from "@babel/runtime-corejs3/core-js-stable/object/assign";
import _includesInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/includes";
import BaseFoundation from '../base/foundation';

class CheckboxFoundation extends BaseFoundation {
  constructor(adapter) {
    super(_Object$assign({}, adapter));
  } // eslint-disable-next-line @typescript-eslint/no-empty-function


  init() {}

  getEvent(checked, e) {
    const props = this.getProps();
    const cbValue = {
      target: _Object$assign(_Object$assign({}, props), {
        checked
      }),
      stopPropagation: () => {
        e.stopPropagation();
      },
      preventDefault: () => {
        e.preventDefault();
      },
      nativeEvent: {
        stopImmediatePropagation: () => {
          if (e.nativeEvent && typeof e.nativeEvent.stopImmediatePropagation === 'function') {
            e.nativeEvent.stopImmediatePropagation();
          }
        }
      }
    };
    return cbValue;
  }

  notifyChange(checked, e) {
    const cbValue = this.getEvent(checked, e);

    this._adapter.notifyChange(cbValue);
  }

  handleChange(e) {
    const disabled = this.getProp('disabled');

    if (disabled) {
      return;
    }

    const isInGroup = this._adapter.getIsInGroup();

    if (isInGroup) {
      const groupDisabled = this._adapter.getGroupDisabled();

      if (!groupDisabled) {
        this.handleChangeInGroup(e);
      }

      return;
    }

    const checked = this.getState('checked');
    const newChecked = !checked;

    if (this._isControlledComponent('checked')) {
      this.notifyChange(newChecked, e);
    } else {
      this.setChecked(newChecked);
      this.notifyChange(newChecked, e);
    }
  }

  handleChangeInGroup(e) {
    const {
      value
    } = this.getProps();

    const groupValue = this._adapter.getGroupValue();

    const checked = _includesInstanceProperty(groupValue).call(groupValue, value);

    const newChecked = !checked;
    const event = this.getEvent(newChecked, e);

    this._adapter.notifyChange(event);

    this._adapter.notifyGroupChange(event);
  }

  setChecked(checked) {
    this._adapter.setNativeControlChecked(checked);
  } // eslint-disable-next-line @typescript-eslint/no-empty-function


  destroy() {}

}

export default CheckboxFoundation;
