// @builder/mpa-config/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateMPAEntries = void 0;
var path = require("path");
var app_helpers_1 = require("@builder/app-helpers");
var generate_1 = require("./generate");
var generateMPAEntries = function (api, options) {
    var context = api.context;
    var _a = options.framework, framework = _a === void 0 ? 'rax' : _a, _b = options.targetDir, targetDir = _b === void 0 ? '' : _b;
    var entries = options.entries;
    var rootDir = context.rootDir, commandArgs = context.commandArgs;
    if (commandArgs.mpaEntry) {
        var arr_1 = commandArgs.mpaEntry.split(',');
        entries = entries.filter(function (entry) {
            return arr_1.includes(entry.entryName);
        });
    }
    var parsedEntries = {};
    entries.forEach(function (entry) {
        var entryName = entry.entryName, entryPath = entry.entryPath, pageConfig = __rest(entry, ["entryName", "entryPath"]);
        var source = pageConfig.source;
        var isAppEntry = /app(\.(t|j)sx?)?$/.test(entryPath);
        // when the entry has no export default declaration, do not generate any files
        var finalEntry = entryPath;
        var runAppPath = null;
        var routesFilePath;
        if (isAppEntry || (0, app_helpers_1.checkExportDefaultDeclarationExists)(path.join(rootDir, 'src', source))) {
            var result = (0, generate_1.default)(api, { framework: framework, targetDir: targetDir, pageEntry: entryPath, entryName: entryName, pageConfig: pageConfig, isAppEntry: isAppEntry });
            finalEntry = result.entryPath;
            runAppPath = result.runAppPath;
            routesFilePath = result.routesFilePath;
        }
        parsedEntries[entryName] = __assign(__assign({}, entry), { finalEntry: finalEntry, shouldRedirectRunApp: isAppEntry, runAppPath: runAppPath, routesFilePath: routesFilePath });
    });
    return parsedEntries;
};
exports.generateMPAEntries = generateMPAEntries;
var setMPAConfig = function (api, config, options) {
    if (!options) {
        throw new Error('There need pass options param to setMPAConfig method');
    }
    var parsedEntries = (0, exports.generateMPAEntries)(api, options);
    // clear entry points
    config.entryPoints.clear();
    // add mpa entries
    var matchStrs = [];
    // add redirect entry path
    var redirectEntries = [];
    Object.keys(parsedEntries).forEach(function (entryKey) {
        var _a = parsedEntries[entryKey], entryName = _a.entryName, finalEntry = _a.finalEntry, shouldRedirectRunApp = _a.shouldRedirectRunApp, runAppPath = _a.runAppPath, routesFilePath = _a.routesFilePath;
        config.entry(entryName).add(finalEntry);
        if (shouldRedirectRunApp) {
            redirectEntries.push({
                entryPath: finalEntry,
                runAppPath: runAppPath,
                routesFilePath: routesFilePath,
            });
        }
        // get page paths for rule match
        matchStrs.push((0, app_helpers_1.formatPath)(routesFilePath));
    });
    api.applyMethod('addImportDeclaration', {
        multipleSource: {
            runApp: redirectEntries.map(function (_a) {
                var entryPath = _a.entryPath, runAppPath = _a.runAppPath;
                return ({
                    filename: entryPath,
                    value: runAppPath,
                    type: 'normal',
                });
            }),
        },
    });
    if (config.plugins.has('document')) {
        config.plugin('document').tap(function (args) {
            return [__assign(__assign({}, args[0]), { pages: Object.values(parsedEntries) })];
        });
    }
    // modify appJSON rules for mpa
    var matchInclude = function (filepath) {
        var matchReg = matchStrs.length ? new RegExp(matchStrs.join('|')) : null;
        return matchReg && matchReg.test(filepath);
    };
    config.module.rule('appJSON').include.add(matchInclude);
};
exports.default = setMPAConfig;
