// build-plugin-ice-store/lib/replacePathLoader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var fse = require("fs-extra");
var app_helpers_1 = require("@builder/app-helpers");
var loader_utils_1 = require("loader-utils");
var LAYOUT_DIR = 'Layout';
function default_1(content, sourceMap) {
    // async loader
    var callback = this.async();
    this.cacheable();
    loader.call(this, content, sourceMap).then(function (args) { return callback.apply(void 0, __spreadArray([null], args, false)); }, function (err) { return callback(err); });
}
exports.default = default_1;
function loader(content, sourceMap) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, alias, tempDir, srcPath, rootDir, routesPaths, applyMethod, currentRoutesPath, pagesName, _loop_1, _i, pagesName_1, pageName;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = (0, loader_utils_1.getOptions)(this), alias = _a.alias, tempDir = _a.tempDir, srcPath = _a.srcPath, rootDir = _a.rootDir, routesPaths = _a.routesPaths, applyMethod = _a.applyMethod;
                    currentRoutesPath = this.resourcePath;
                    if (!routesPaths.includes(currentRoutesPath)) {
                        // if this.resourcePath is htmlWebpackPluginPublicPath, skip it
                        return [2 /*return*/, [content, sourceMap]];
                    }
                    pagesName = applyMethod('getPages', srcPath);
                    _loop_1 = function (pageName) {
                        var absolutePagePath, absolutePageLayoutPath, source;
                        return __generator(this, function (_c) {
                            switch (_c.label) {
                                case 0:
                                    absolutePagePath = path.join(rootDir, 'src', 'pages', pageName);
                                    absolutePageLayoutPath = path.join(absolutePagePath, LAYOUT_DIR);
                                    source = function (originSourcePath) {
                                        var originSourceBasename = path.basename(originSourcePath);
                                        if (['index', 'index.js', 'index.jsx', 'index.tsx'].includes(originSourceBasename)) {
                                            // ./pages/Home/index[.js|.jsx|.tsx] -> ./pages/Home
                                            originSourcePath = path.dirname(originSourcePath);
                                        }
                                        if (!Object.keys(alias).length) {
                                            alias['@'] = 'src';
                                        }
                                        var matchAliasKey = '';
                                        Object.keys(alias).forEach(function (aliasKey) {
                                            if (new RegExp(aliasKey).test(originSourcePath)) {
                                                matchAliasKey = aliasKey;
                                            }
                                        });
                                        var absoluteSourcePath;
                                        if (matchAliasKey) {
                                            // handle alias path
                                            var matchAliasPath = alias[matchAliasKey]; // .src/*
                                            var replaceWithAliasPath = originSourcePath.replace(RegExp(matchAliasKey), matchAliasPath.replace(/\*/g, ''));
                                            absoluteSourcePath = path.join(rootDir, replaceWithAliasPath);
                                        }
                                        else {
                                            // handle relative path
                                            var currentRoutesDir = path.dirname(currentRoutesPath);
                                            absoluteSourcePath = path.join(currentRoutesDir, originSourcePath);
                                        }
                                        return [absolutePagePath, absolutePageLayoutPath].includes(absoluteSourcePath);
                                    };
                                    return [4 /*yield*/, (0, app_helpers_1.redirectImport)(content, {
                                            source: source,
                                            redirectImports: [
                                                {
                                                    redirectPath: generateRedirectPath({ tempDir: tempDir, pageName: pageName, rootDir: rootDir }),
                                                    default: true,
                                                }
                                            ]
                                        })];
                                case 1:
                                    // eslint-disable-next-line no-await-in-loop
                                    content = _c.sent();
                                    return [2 /*return*/];
                            }
                        });
                    };
                    _i = 0, pagesName_1 = pagesName;
                    _b.label = 1;
                case 1:
                    if (!(_i < pagesName_1.length)) return [3 /*break*/, 4];
                    pageName = pagesName_1[_i];
                    return [5 /*yield**/, _loop_1(pageName)];
                case 2:
                    _b.sent();
                    _b.label = 3;
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/, [content, sourceMap]];
            }
        });
    });
}
function generateRedirectPath(_a) {
    var tempDir = _a.tempDir, pageName = _a.pageName, rootDir = _a.rootDir;
    if (fse.pathExistsSync(path.join(rootDir, 'src', 'pages', pageName, LAYOUT_DIR))) {
        return (0, app_helpers_1.formatPath)(path.join(rootDir, tempDir, 'pages', pageName, LAYOUT_DIR));
    }
    var pagePath = path.join(rootDir, tempDir, 'pages', pageName, 'index.tsx');
    return (0, app_helpers_1.formatPath)(pagePath);
}
