// build-plugin-react-app/lib/validateViteConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("@builder/pack/deps/lodash");
var invalidConfig = {
    root: false,
    mode: false,
    base: 'publicPath',
    define: 'define',
    plugins: 'vitePlugins',
    publicDir: false,
    'resolve.alias': 'alias',
    'css.postcss': 'postcssOptions',
    'css.preprocessorOptions.scss': 'sassLoaderOptions',
    'css.preprocessorOptions.less': 'lessLoaderOptions',
    'server.host': '--host',
    'server.post': '--post',
    'server.https': '--https',
    'build.outDir': 'outputDir',
    'build.minify': 'minify',
    'build.terserOptions': 'minify',
};
var getInvalidMessage = function (userConfig) {
    var msg = '';
    Object.keys(invalidConfig).some(function (configKey) {
        var viteConfigKey = "vite." + configKey;
        if (_.get(userConfig, viteConfigKey) !== undefined) {
            msg = "configuration of " + viteConfigKey + " is invalid in icejs";
            if (invalidConfig[configKey]) {
                msg += ", try to use " + invalidConfig[configKey];
            }
            return true;
        }
        return false;
    });
    return msg;
};
exports.default = getInvalidMessage;
