// @unocss/preset-uno/dist/rules.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const _default = require('./chunks/default.cjs');
require('@unocss/core');
require('./chunks/index.cjs');
require('./chunks/pseudo.cjs');



exports.alignContents = _default.alignContents;
exports.alignItems = _default.alignItems;
exports.alignSelfs = _default.alignSelfs;
exports.animations = _default.animations;
exports.appearance = _default.appearance;
exports.appearances = _default.appearances;
exports.aspectRatio = _default.aspectRatio;
exports.bgAttachments = _default.bgAttachments;
exports.bgBlendModes = _default.bgBlendModes;
exports.bgClips = _default.bgClips;
exports.bgColors = _default.bgColors;
exports.bgGradients = _default.bgGradients;
exports.bgImages = _default.bgImages;
exports.bgOrigins = _default.bgOrigins;
exports.bgPositions = _default.bgPositions;
exports.bgRepeats = _default.bgRepeats;
exports.bgSizes = _default.bgSizes;
exports.borderColors = _default.borderColors;
exports.borderRadius = _default.borderRadius;
exports.borderSizes = _default.borderSizes;
exports.borderStyles = _default.borderStyles;
exports.borders = _default.borders;
exports.boxDecorationBreaks = _default.boxDecorationBreaks;
exports.boxShadows = _default.boxShadows;
exports.boxSizing = _default.boxSizing;
exports.breaks = _default.breaks;
exports.caretColors = _default.caretColors;
exports.caretOpacity = _default.caretOpacity;
exports.colorResolver = _default.colorResolver;
exports.container = _default.container;
exports.containerShortcuts = _default.containerShortcuts;
exports.contents = _default.contents;
exports.cssVariables = _default.cssVariables;
exports.cursors = _default.cursors;
exports.displays = _default.displays;
exports.divideColors = _default.divideColors;
exports.divideSizes = _default.divideSizes;
exports.divideStyles = _default.divideStyles;
exports.divides = _default.divides;
exports.fillColors = _default.fillColors;
exports.filters = _default.filters;
exports.flex = _default.flex;
exports.floats = _default.floats;
exports.fontSizes = _default.fontSizes;
exports.fontSmoothings = _default.fontSmoothings;
exports.fontStyles = _default.fontStyles;
exports.fontVariantNumeric = _default.fontVariantNumeric;
exports.fontWeights = _default.fontWeights;
exports.fonts = _default.fonts;
exports.fontsFamilies = _default.fontsFamilies;
exports.gaps = _default.gaps;
exports.grids = _default.grids;
exports.hyphens = _default.hyphens;
exports.imageRenderings = _default.imageRenderings;
exports.insets = _default.insets;
exports.isolations = _default.isolations;
exports.justifies = _default.justifies;
exports.justifyItems = _default.justifyItems;
exports.justifySelfs = _default.justifySelfs;
exports.leadings = _default.leadings;
exports.lineClamps = _default.lineClamps;
exports.listStyle = _default.listStyle;
exports.margins = _default.margins;
exports.mixBlendModes = _default.mixBlendModes;
exports.objectPositions = _default.objectPositions;
exports.opacity = _default.opacity;
exports.orders = _default.orders;
exports.outline = _default.outline;
exports.overflows = _default.overflows;
exports.overscrolls = _default.overscrolls;
exports.paddings = _default.paddings;
exports.parseColorUtil = _default.parseColorUtil;
exports.placeContents = _default.placeContents;
exports.placeItems = _default.placeItems;
exports.placeSelfs = _default.placeSelfs;
exports.placeholder = _default.placeholder;
exports.pointerEvents = _default.pointerEvents;
exports.positions = _default.positions;
exports.questionMark = _default.questionMark;
exports.resizes = _default.resizes;
exports.ringColors = _default.ringColors;
exports.ringOffsetColors = _default.ringOffsetColors;
exports.rings = _default.rings;
exports.rules = _default.rules;
exports.screenReadersAccess = _default.screenReadersAccess;
exports.sizes = _default.sizes;
exports.tabSizes = _default.tabSizes;
exports.tables = _default.tables;
exports.textAligns = _default.textAligns;
exports.textColors = _default.textColors;
exports.textDecorationColors = _default.textDecorationColors;
exports.textDecorationLengths = _default.textDecorationLengths;
exports.textDecorationOffsets = _default.textDecorationOffsets;
exports.textDecorationStyles = _default.textDecorationStyles;
exports.textDecorations = _default.textDecorations;
exports.textIndents = _default.textIndents;
exports.textOverflows = _default.textOverflows;
exports.textShadows = _default.textShadows;
exports.textStrokeColors = _default.textStrokeColors;
exports.textStrokeWidths = _default.textStrokeWidths;
exports.textTransforms = _default.textTransforms;
exports.trackings = _default.trackings;
exports.transforms = _default.transforms;
exports.transitions = _default.transitions;
exports.userSelects = _default.userSelects;
exports.varEmpty = _default.varEmpty;
exports.verticalAligns = _default.verticalAligns;
exports.whitespaces = _default.whitespaces;
exports.wordSpacings = _default.wordSpacings;
exports.writingModes = _default.writingModes;
exports.writingOrientations = _default.writingOrientations;
exports.zIndexes = _default.zIndexes;
