// stylelint-config-ali/essential.js
/*
 * 此文件内的规则作为「必修问题」
 */

'use strict';

const miniappTags = require('./miniapp-tags');

module.exports = {
  plugins: ['stylelint-scss'],
  rules: {
    /*
     * stylelint possible-errors 类目下的规则，即 stylelint-config-recommended 中的规则
     * https://stylelint.io/user-guide/rules/#possible-errors
     */
    'at-rule-no-unknown': null,
    'scss/at-rule-no-unknown': true,
    'block-no-empty': null, // 不作为必修问题
    'color-no-invalid-hex': true,
    'comment-no-empty': true,
    'declaration-block-no-duplicate-properties': [
      true,
      {
        ignore: ['consecutive-duplicates-with-different-values'],
      },
    ],
    'declaration-block-no-shorthand-property-overrides': true,
    'font-family-no-duplicate-names': true,
    // 'font-family-no-missing-generic-family-keyword': true, // 关闭此条规则：使用 iconfont 字体时会报错
    'function-calc-no-unspaced-operator': true,
    'function-linear-gradient-no-nonstandard-direction': true,
    'keyframe-declaration-no-important': null, // 不作为必修问题
    'media-feature-name-no-unknown': true,
    'no-descending-specificity': null, // 此条规则关闭：实际中有很多这样用的，且多数人熟悉css优先级
    'no-duplicate-at-import-rules': true,
    'no-duplicate-selectors': true,
    'no-empty-source': true,
    'no-extra-semicolons': true,
    'no-invalid-double-slash-comments': true,
    'property-no-unknown': true,
    'selector-pseudo-class-no-unknown': [
      true,
      {
        ignorePseudoClasses: ['global', 'local', 'export'],
      },
    ],
    'selector-pseudo-element-no-unknown': true,
    'selector-type-no-unknown': [
      true,
      {
        ignore: ['custom-elements'],
        ignoreTypes: miniappTags,
      },
    ],
    'string-no-newline': true,
    'unit-no-unknown': [
      true,
      {
        ignoreUnits: ['rpx'],
      },
    ],

    /*
     * 集团 CSS 规约中「强制」级别的规则
     */
    'declaration-block-trailing-semicolon': 'always',
  },
  ignoreFiles: ['**/*.js', '**/*.jsx', '**/*.ts', '**/*.tsx'],
};
