// @unocss/preset-uno/dist/chunks/default.mjs
import { hex2rgba, toArray } from '@unocss/core';
import { h as handler, d as directionMap, c as cornerMap, a as capitalize, x as xyzMap } from './index.mjs';
import { C as CONTROL_BYPASS_PSEUDO } from './pseudo.mjs';

const verticalAlignAlias = {
  mid: "middle",
  base: "baseline",
  btm: "bottom"
};
const verticalAligns = [
  [/^(?:vertical|align|v)-(baseline|top|bottom|middle|text-top|text-bottom|mid|base|btm)$/, ([, v]) => ({ "vertical-align": verticalAlignAlias[v] || v })]
];
const textAligns = [
  ["text-center", { "text-align": "center" }],
  ["text-left", { "text-align": "left" }],
  ["text-right", { "text-align": "right" }],
  ["text-justify", { "text-align": "justify" }]
];

const keyframes = {
  "spin": "{from{transform:rotate(0deg)}to{transform:rotate(360deg)}}",
  "ping": "{0%{transform:scale(1);opacity:1}75%,100%{transform:scale(2);opacity:0}}",
  "bounce": "{from,20%,53%,80%,to{animation-timing-function:cubic-bezier(0.215,0.61,0.355,1);transform:translate3d(0,0,0)}40%,43%{animation-timing-function:cubic-bezier(0.755,0.05,0.855,0.06);transform:translate3d(0,-30px,0)}70%{animation-timing-function:cubic-bezier(0.755,0.05,0.855,0.06);transform:translate3d(0,-15px,0)}90%{transform:translate3d(0,-4px,0)}}",
  "flash": "{from,50%,to{opacity:1}25%,75%{opacity:0}}",
  "pulse": "{from{transform:scale3d(1,1,1)}50%{transform:scale3d(1.05,1.05,1.05)}to{transform:scale3d(1,1,1)}}",
  "rubber-band": "{from{transform:scale3d(1,1,1)}30%{transform:scale3d(1.25,0.75,1)}40%{transform:scale3d(0.75,1.25,1)}50%{transform:scale3d(1.15,0.85,1)}65%{transform:scale3d(0.95,1.05,1)}75%{transform:scale3d(1.05,0.95,1)}to{transform:scale3d(1,1,1)}}",
  "shake-x": "{from,to{transform:translate3d(0,0,0)}10%,30%,50%,70%,90%{transform:translate3d(-10px,0,0)}20%,40%,60%,80%{transform:translate3d(10px,0,0)}}",
  "shake-y": "{from,to{transform:translate3d(0,0,0)}10%,30%,50%,70%,90%{transform:translate3d(0,-10px,0)}20%,40%,60%,80%{transform:translate3d(0,10px,0)}}",
  "head-shake": "{0%{transform:translateX(0)}6.5%{transform:translateX(-6px) rotateY(-9deg)}18.5%{transform:translateX(5px) rotateY(7deg)}31.5%{transform:translateX(-3px) rotateY(-5deg)}43.5%{transform:translateX(2px) rotateY(3deg)}50%{transform:translateX(0)}}",
  "swing": "{20%{transform:rotate3d(0,0,1,15deg)}40%{transform:rotate3d(0,0,1,-10deg)}60%{transform:rotate3d(0,0,1,5deg)}80%{transform:rotate3d(0,0,1,-5deg)}to{transform:rotate3d(0,0,1,0deg)}}",
  "tada": "{from{transform:scale3d(1,1,1)}10%,20%{transform:scale3d(0.9,0.9,0.9) rotate3d(0,0,1,-3deg)}30%,50%,70%,90%{transform:scale3d(1.1,1.1,1.1) rotate3d(0,0,1,3deg)}40%,60%,80%{transform:scale3d(1.1,1.1,1.1) rotate3d(0,0,1,-3deg)}to{transform:scale3d(1,1,1)}}",
  "wobble": "{from{transform:translate3d(0,0,0)}15%{transform:translate3d(-25%,0,0) rotate3d(0,0,1,-5deg)}30%{transform:translate3d(20%,0,0) rotate3d(0,0,1,3deg)}45%{transform:translate3d(-15%,0,0) rotate3d(0,0,1,-3deg)}60%{transform:translate3d(10%,0,0) rotate3d(0,0,1,2deg)}75%{transform:translate3d(-5%,0,0) rotate3d(0,0,1,-1deg)}to{transform:translate3d(0,0,0)}}",
  "jello": "{from,11.1% to{transform:translate3d(0,0,0)}22.2%{transform:skewX(-12.5deg) skewY(-12.5deg)}33.3%{transform:skewX(6.25deg) skewY(6.25deg)}44.4%{transform:skewX(-3.125deg)skewY(-3.125deg)}55.5%{transform:skewX(1.5625deg) skewY(1.5625deg)}66.6%{transform:skewX(-0.78125deg) skewY(-0.78125deg)}77.7%{transform:skewX(0.390625deg) skewY(0.390625deg)}88.8%{transform:skewX(-0.1953125deg) skewY(-0.1953125deg)}}",
  "heart-beat": "{0%{transform:scale(1)}14%{transform:scale(1.3)}28%{transform:scale(1)}42%{transform:scale(1.3)}70%{transform:scale(1)}}",
  "hinge": "{0%{transform-origin:top left;animation-timing-function:ease-in-out}20%,60%{transform:rotate3d(0,0,1,80deg);transform-origin:top left;animation-timing-function:ease-in-out}40%,80%{transform:rotate3d(0,0,1,60deg);transform-origin:top left;animation-timing-function:ease-in-out}to{transform:translate3d(0,700px,0);opacity:0}}",
  "jack-in-the-box": "{from{opacity:0;transform-origin:center bottom;transform:scale(0.1) rotate(30deg)}50%{transform:rotate(-10deg)}70%{transform:rotate(3deg)}to{transform:scale(1)}}",
  "light-speed-in-left": "{from{opacity:0;transform:translate3d(100%,0,0) skewX(-30deg)}60%{opacity:1;transform:skewX(20deg)}80%{transform:skewX(-5deg)}to{transform:translate3d(0,0,0)}}",
  "light-speed-in-right": "{from{opacity:0;transform:translate3d(100%,0,0) skewX(-30deg)}60%{opacity:1;transform:skewX(20deg)}80%{transform:skewX(-5deg)}to{transform:translate3d(0,0,0)}}",
  "light-speed-out-left": "{from{opacity:1}to{opacity:0;transform:translate3d(100%,0,0) skewX(30deg)}}",
  "light-speed-out-right": "{from{opacity:1}to{opacity:0;transform:translate3d(100%,0,0) skewX(30deg)}}",
  "flip": "{from{transform:perspective(400px) scale3d(1,1,1) translate3d(0,0,0) rotate3d(0,1,0,-360deg);animation-timing-function:ease-out}40%{transform:perspective(400px) scale3d(1,1,1) translate3d(0,0,150px) rotate3d(0,1,0,-190deg);animation-timing-function:ease-out}50%{transform:perspective(400px) scale3d(1,1,1) translate3d(0,0,150px) rotate3d(0,1,0,-170deg);animation-timing-function:ease-in}80%{transform:perspective(400px) scale3d(0.95,0.95,0.95) translate3d(0,0,0) rotate3d(0,1,0,0deg);animation-timing-function:ease-in}to{transform:perspective(400px) scale3d(1,1,1) translate3d(0,0,0) rotate3d(0,1,0,0deg);animation-timing-function:ease-in}}",
  "flip-in-x": "{from{transform:perspective(400px) rotate3d(1,0,0,90deg);animation-timing-function:ease-in;opacity:0}40%{transform:perspective(400px) rotate3d(1,0,0,-20deg);animation-timing-function:ease-in}60%{transform:perspective(400px) rotate3d(1,0,0,10deg);opacity:1}80%{transform:perspective(400px) rotate3d(1,0,0,-5deg)}to{transform:perspective(400px)}}",
  "flip-in-y": "{from{transform:perspective(400px) rotate3d(0,1,0,90deg);animation-timing-function:ease-in;opacity:0}40%{transform:perspective(400px) rotate3d(0,1,0,-20deg);animation-timing-function:ease-in}60%{transform:perspective(400px) rotate3d(0,1,0,10deg);opacity:1}80%{transform:perspective(400px) rotate3d(0,1,0,-5deg)}to{transform:perspective(400px)}}",
  "flip-out-x": "{from{transform:perspective(400px)}30%{transform:perspective(400px) rotate3d(1,0,0,-20deg);opacity:1}to{transform:perspective(400px) rotate3d(1,0,0,90deg);opacity:0}}",
  "flip-out-y": "{from{transform:perspective(400px)}30%{transform:perspective(400px) rotate3d(0,1,0,-15deg);opacity:1}to{transform:perspective(400px) rotate3d(0,1,0,90deg);opacity:0}}",
  "rotate-in": "{from{transform-origin:center;transform:rotate3d(0,0,1,-200deg);opacity:0}to{transform-origin:center;transform:translate3d(0,0,0);opacity:1}}",
  "rotate-in-down-left": "{from{transform-origin:left bottom;transform:rotate3d(0,0,1,-45deg);opacity:0}to{transform-origin:left bottom;transform:translate3d(0,0,0);opacity:1}}",
  "rotate-in-down-right": "{from{transform-origin:right bottom;transform:rotate3d(0,0,1,45deg);opacity:0}to{transform-origin:right bottom;transform:translate3d(0,0,0);opacity:1}}",
  "rotate-in-up-left": "{from{transform-origin:left top;transform:rotate3d(0,0,1,45deg);opacity:0}to{transform-origin:left top;transform:translate3d(0,0,0);opacity:1}}",
  "rotate-in-up-right": "{from{transform-origin:right bottom;transform:rotate3d(0,0,1,-90deg);opacity:0}to{transform-origin:right bottom;transform:translate3d(0,0,0);opacity:1}}",
  "rotate-out": "{from{transform-origin:center;opacity:1}to{transform-origin:center;transform:rotate3d(0,0,1,200deg);opacity:0}}",
  "rotate-out-down-left": "{from{transform-origin:left bottom;opacity:1}to{transform-origin:left bottom;transform:rotate3d(0,0,1,45deg);opacity:0}}",
  "rotate-out-down-right": "{from{transform-origin:right bottom;opacity:1}to{transform-origin:right bottom;transform:rotate3d(0,0,1,-45deg);opacity:0}}",
  "rotate-out-up-left": "{from{transform-origin:left bottom;opacity:1}to{transform-origin:left bottom;transform:rotate3d(0,0,1,-45deg);opacity:0}}",
  "rotate-out-up-right": "{from{transform-origin:right bottom;opacity:1}to{transform-origin:left bottom;transform:rotate3d(0,0,1,90deg);opacity:0}}",
  "roll-in": "{from{opacity:0;transform:translate3d(-100%,0,0) rotate3d(0,0,1,-120deg)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "roll-out": "{from{opacity:1}to{opacity:0;transform:translate3d(100%,0,0) rotate3d(0,0,1,120deg)}}",
  "zoom-in": "{from{opacity:0;transform:scale3d(0.3,0.3,0.3)}50%{opacity:1}}",
  "zoom-in-down": "{from{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(0,-1000px,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}60%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(0,60px,0);animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "zoom-in-left": "{from{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(-1000px,0,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}60%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(10px,0,0);animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "zoom-in-right": "{from{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(1000px,0,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}60%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(-10px,0,0);animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "zoom-in-up": "{from{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(0,1000px,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}60%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(0,-60px,0);animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "zoom-out": "{from{opacity:1}50%{opacity:0;transform:scale3d(0.3,0.3,0.3)}to{opacity:0}}",
  "zoom-out-down": "{40%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(0,-60px,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}to{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(0,2000px,0);transform-origin:center bottom;animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "zoom-out-left": "{40%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(42px,0,0)}to{opacity:0;transform:scale(0.1) translate3d(-2000px,0,0);transform-origin:left center}}",
  "zoom-out-right": "{40%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(-42px,0,0)}to{opacity:0;transform:scale(0.1) translate3d(2000px,0,0);transform-origin:right center}}",
  "zoom-out-up": "{40%{opacity:1;transform:scale3d(0.475,0.475,0.475) translate3d(0,60px,0);animation-timing-function:cubic-bezier(0.55,0.055,0.675,0.19)}to{opacity:0;transform:scale3d(0.1,0.1,0.1) translate3d(0,-2000px,0);transform-origin:center bottom;animation-timing-function:cubic-bezier(0.175,0.885,0.32,1)}}",
  "bounce-in": "{from,20%,40%,60%,80%,to{animation-timing-function:ease-in-out}0%{opacity:0;transform:scale3d(0.3,0.3,0.3)}20%{transform:scale3d(1.1,1.1,1.1)}40%{transform:scale3d(0.9,0.9,0.9)}60%{transform:scale3d(1.03,1.03,1.03);opacity:1}80%{transform:scale3d(0.97,0.97,0.97)}to{opacity:1;transform:scale3d(1,1,1)}}",
  "bounce-in-down": "{from,60%,75%,90%,to{animation-timing-function:cubic-bezier(0.215,0.61,0.355,1)}0%{opacity:0;transform:translate3d(0,-3000px,0)}60%{opacity:1;transform:translate3d(0,25px,0)}75%{transform:translate3d(0,-10px,0)}90%{transform:translate3d(0,5px,0)}to{transform:translate3d(0,0,0)}}",
  "bounce-in-left": "{from,60%,75%,90%,to{animation-timing-function:cubic-bezier(0.215,0.61,0.355,1)}0%{opacity:0;transform:translate3d(-3000px,0,0)}60%{opacity:1;transform:translate3d(25px,0,0)}75%{transform:translate3d(-10px,0,0)}90%{transform:translate3d(5px,0,0)}to{transform:translate3d(0,0,0)}}",
  "bounce-in-right": "{from,60%,75%,90%,to{animation-timing-function:cubic-bezier(0.215,0.61,0.355,1)}0%{opacity:0;transform:translate3d(3000px,0,0)}60%{opacity:1;transform:translate3d(-25px,0,0)}75%{transform:translate3d(10px,0,0)}90%{transform:translate3d(-5px,0,0)}to{transform:translate3d(0,0,0)}}",
  "bounce-in-up": "{from,60%,75%,90%,to{animation-timing-function:cubic-bezier(0.215,0.61,0.355,1)}0%{opacity:0;transform:translate3d(0,3000px,0)}60%{opacity:1;transform:translate3d(0,-20px,0)}75%{transform:translate3d(0,10px,0)}90%{transform:translate3d(0,-5px,0)}to{transform:translate3d(0,0,0)}}",
  "bounce-out": "{20%{transform:scale3d(0.9,0.9,0.9)}50%,55%{opacity:1;transform:scale3d(1.1,1.1,1.1)}to{opacity:0;transform:scale3d(0.3,0.3,0.3)}}",
  "bounce-out-down": "{20%{transform:translate3d(0,10px,0)}40%,45%{opacity:1;transform:translate3d(0,-20px,0)}to{opacity:0;transform:translate3d(0,2000px,0)}}",
  "bounce-out-left": "{20%{opacity:1;transform:translate3d(20px,0,0)}to{opacity:0;transform:translate3d(-2000px,0,0)}}",
  "bounce-out-right": "{20%{opacity:1;transform:translate3d(-20px,0,0)}to{opacity:0;transform:translate3d(2000px,0,0)}}",
  "bounce-out-up": "{20%{transform:translate3d(0,-10px,0)}40%,45%{opacity:1;transform:translate3d(0,20px,0)}to{opacity:0;transform:translate3d(0,-2000px,0)}}",
  "slide-in-down": "{from{transform:translate3d(0,-100%,0);visibility:visible}to{transform:translate3d(0,0,0)}}",
  "slide-in-left": "{from{transform:translate3d(-100%,0,0);visibility:visible}to{transform:translate3d(0,0,0)}}",
  "slide-in-right": "{from{transform:translate3d(100%,0,0);visibility:visible}to{transform:translate3d(0,0,0)}}",
  "slide-in-up": "{from{transform:translate3d(0,100%,0);visibility:visible}to{transform:translate3d(0,0,0)}}",
  "slide-out-down": "{from{transform:translate3d(0,0,0)}to{visibility:hidden;transform:translate3d(0,100%,0)}}",
  "slide-out-left": "{from{transform:translate3d(0,0,0)}to{visibility:hidden;transform:translate3d(-100%,0,0)}}",
  "slide-out-right": "{from{transform:translate3d(0,0,0)}to{visibility:hidden;transform:translate3d(100%,0,0)}}",
  "slide-out-up": "{from{transform:translate3d(0,0,0)}to{visibility:hidden;transform:translate3d(0,-100%,0)}}",
  "fade-in": "{from{opacity:0}to{opacity:1}}",
  "fade-in-down": "{from{opacity:0;transform:translate3d(0,-100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-down-big": "{from{opacity:0;transform:translate3d(0,-2000px,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-left": "{from{opacity:0;transform:translate3d(-100%,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-left-big": "{from{opacity:0;transform:translate3d(-2000px,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-right": "{from{opacity:0;transform:translate3d(100%,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-right-big": "{from{opacity:0;transform:translate3d(2000px,0,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-up": "{from{opacity:0;transform:translate3d(0,100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-up-big": "{from{opacity:0;transform:translate3d(0,2000px,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-top-left": "{from{opacity:0;transform:translate3d(-100%,-100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-top-right": "{from{opacity:0;transform:translate3d(100%,-100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-bottom-left": "{from{opacity:0;transform:translate3d(-100%,100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-in-bottom-right": "{from{opacity:0;transform:translate3d(100%,100%,0)}to{opacity:1;transform:translate3d(0,0,0)}}",
  "fade-out": "{from{opacity:1}to{opacity:0}}",
  "fade-out-down": "{from{opacity:1}to{opacity:0;transform:translate3d(0,100%,0)}}",
  "fade-out-down-big": "{from{opacity:1}to{opacity:0;transform:translate3d(0,2000px,0)}}",
  "fade-out-left": "{from{opacity:1}to{opacity:0;transform:translate3d(-100%,0,0)}}",
  "fade-out-left-big": "{from{opacity:1}to{opacity:0;transform:translate3d(-2000px,0,0)}}",
  "fade-out-right": "{from{opacity:1}to{opacity:0;transform:translate3d(100%,0,0)}}",
  "fade-out-right-big": "{from{opacity:1}to{opacity:0;transform:translate3d(2000px,0,0)}}",
  "fade-out-up": "{from{opacity:1}to{opacity:0;transform:translate3d(0,-100%,0)}}",
  "fade-out-up-big": "{from{opacity:1}to{opacity:0;transform:translate3d(0,-2000px,0)}}",
  "fade-out-top-left": "{from{opacity:1;transform:translate3d(0,0,0)}to{opacity:0;transform:translate3d(-100%,-100%,0)}}",
  "fade-out-top-right": "{from{opacity:1;transform:translate3d(0,0,0)}to{opacity:0;transform:translate3d(100%,-100%,0)}}",
  "fade-out-bottom-left": "{from{opacity:1;transform:translate3d(0,0,0)}to{opacity:0;transform:translate3d(-100%,100%,0)}}",
  "fade-out-bottom-right": "{from{opacity:1;transform:translate3d(0,0,0)}to{opacity:0;transform:translate3d(100%,100%,0)}}",
  "back-in-up": "{0%{opacity:0.7;transform:translateY(1200px) scale(0.7)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:1;transform:scale(1)}}",
  "back-in-down": "{0%{opacity:0.7;transform:translateY(-1200px) scale(0.7)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:1;transform:scale(1)}}",
  "back-in-right": "{0%{opacity:0.7;transform:translateX(2000px) scale(0.7)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:1;transform:scale(1)}}",
  "back-in-left": "{0%{opacity:0.7;transform:translateX(-2000px) scale(0.7)}80%{opacity:0.7;transform:translateX(0px) scale(0.7)}100%{opacity:1;transform:scale(1)}}",
  "back-out-up": "{0%{opacity:1;transform:scale(1)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:0.7;transform:translateY(-700px) scale(0.7)}}",
  "back-out-down": "{0%{opacity:1;transform:scale(1)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:0.7;transform:translateY(700px) scale(0.7)}}",
  "back-out-right": "{0%{opacity:1;transform:scale(1)}80%{opacity:0.7;transform:translateY(0px) scale(0.7)}100%{opacity:0.7;transform:translateX(2000px) scale(0.7)}}",
  "back-out-left": "{0%{opacity:1;transform:scale(1)}80%{opacity:0.7;transform:translateX(-2000px) scale(0.7)}100%{opacity:0.7;transform:translateY(-700px) scale(0.7)}}"
};
const durations = {
  "heart-beat": "1.3s",
  "bounce-in": "0.75s",
  "bounce-out": "0.75s",
  "flip-out-x": "0.75s",
  "flip-out-y": "0.75s",
  "hinge": "2s"
};
const timingFns = {
  "head-shake": "ease-in-out",
  "heart-beat": "ease-in-out",
  "pulse": "ease-in-out",
  "light-speed-in-left": "ease-out",
  "light-speed-in-right": "ease-out",
  "light-speed-out-left": "ease-in",
  "light-speed-out-right": "ease-in"
};
const properties = {
  "bounce": { "transform-origin": "center bottom" },
  "jello": { "transform-origin": "center" },
  "swing": { "transform-origin": "top center" },
  "flip": { "backface-visibility": "visible" },
  "flip-in-x": { "backface-visibility": "visible !important" },
  "flip-in-y": { "backface-visibility": "visible !important" },
  "flip-out-x": { "backface-visibility": "visible !important" },
  "flip-out-y": { "backface-visibility": "visible !important" },
  "rotate-in": { "transform-origin": "center" },
  "rotate-in-down-left": { "transform-origin": "left bottom" },
  "rotate-in-down-right": { "transform-origin": "right bottom" },
  "rotate-in-up-left": { "transform-origin": "left bottom" },
  "rotate-in-up-right": { "transform-origin": "right bottom" },
  "rotate-out": { "transform-origin": "center" },
  "rotate-out-down-left": { "transform-origin": "left bottom" },
  "rotate-out-down-right": { "transform-origin": "right bottom" },
  "rotate-out-up-left": { "transform-origin": "left bottom" },
  "rotate-out-up-right": { "transform-origin": "right bottom" },
  "hinge": { "transform-origin": "top left" },
  "zoom-out-down": { "transform-origin": "center bottom" },
  "zoom-out-left": { "transform-origin": "left center" },
  "zoom-out-right": { "transform-origin": "right center" },
  "zoom-out-up": { "transform-origin": "center bottom" }
};
const animations = [
  [/^animate-(.*)$/, ([, name], { constructCSS }) => {
    const kf = keyframes[name];
    if (kf) {
      return `@keyframes ${name}${kf}
${constructCSS(Object.assign({ animation: `${name} ${durations[name] || "1s"} ${timingFns[name] || "linear"} infinite` }, properties[name]))}`;
    }
  }],
  ["animate-none", { animation: "none" }],
  [/^animate(?:-duration)?-((.+)(?:(s|ms)?))$/, ([, d]) => ({ "animation-duration": handler.bracket.time(d.replace(/-duration/, "")) })],
  [/^animate-delay-((.+)(?:(s|ms)?))$/, ([, d]) => ({ "animation-delay": handler.bracket.time(d) })],
  [/^animate-(?:fill-)?mode-(none|forwards|backwards|both|inherit|initial|revert|unset)$/, ([, d]) => ({ "animation-fill-mode": d })],
  [/^animate-(?:direction-)?(normal|reverse|alternate|alternate-reverse|inherit|initial|revert|unset)$/, ([, d]) => ({ "animation-direction": d })],
  [/^animate-(?:iteration-)?count-(.+)$/, ([, d]) => ({ "animation-iteration-count": d.replace(/\-/g, ", ") })],
  [/^animate-name-(.+)/, ([, d]) => ({ "animation-name": d })],
  [/^animate-play(?:-state)?-(paused|running|inherit|initial|revert|unset)$/, ([, d]) => ({ "animation-play-state": d })]
];

const parseColorUtil = (body, theme) => {
  const [main, opacity2] = body.split(/(?:\/|:)/);
  const [name, no = "DEFAULT"] = main.replace(/([a-z])([0-9])/g, "$1-$2").split(/-/g);
  if (!name)
    return;
  let color;
  const bracket = handler.bracket(main) || main;
  if (bracket.startsWith("#"))
    color = bracket.slice(1);
  if (bracket.startsWith("hex-"))
    color = bracket.slice(4);
  if (!color) {
    const colorData = theme.colors?.[name];
    if (typeof colorData === "string")
      color = colorData;
    else if (no && colorData)
      color = colorData[no];
  }
  return {
    opacity: opacity2,
    name,
    no,
    color,
    rgba: hex2rgba(color)
  };
};
const colorResolver$2 = (attribute, varName) => ([, body], { theme }) => {
  const data = parseColorUtil(body, theme);
  if (!data)
    return;
  const { opacity: opacity2, color, rgba } = data;
  if (!color)
    return;
  const a = opacity2 ? opacity2[0] === "[" ? handler.bracket.percent(opacity2) : parseFloat(opacity2) / 100 : rgba?.[3];
  if (rgba) {
    if (a != null && !Number.isNaN(a)) {
      rgba[3] = typeof a === "string" && !a.includes("%") ? parseFloat(a) : a;
      return {
        [attribute]: `rgba(${rgba.join(",")})`
      };
    } else {
      return {
        [`--un-${varName}-opacity`]: 1,
        [attribute]: `rgba(${rgba.slice(0, 3).join(",")},var(--un-${varName}-opacity))`
      };
    }
  } else {
    return {
      [attribute]: color.replace("%alpha", `${a || 1}`)
    };
  }
};
const opacity = [
  [/^op(?:acity)?-?(.+)$/, ([, d]) => ({ opacity: handler.bracket.percent.cssvar(d) })]
];
const textColors = [
  [/^(?:text|color|c)-(.+)$/, colorResolver$2("color", "text")],
  [/^(?:text|color|c)-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-text-opacity": handler.bracket.percent.cssvar(opacity2) })]
];
const textDecorationColors = [
  [/^underline-(.+)$/, (match, ctx) => {
    const result = colorResolver$2("text-decoration-color", "line")(match, ctx);
    if (result) {
      return {
        "-webkit-text-decoration-color": result["text-decoration-color"],
        ...result
      };
    }
  }],
  [/^underline-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-line-opacity": handler.bracket.percent(opacity2) })]
];
const textStrokeColors = [
  [/^text-stroke-(.+)$/, colorResolver$2("-webkit-text-stroke-color", "text-stroke")],
  [/^text-stroke-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-text-stroke-opacity": handler.bracket.percent(opacity2) })]
];
const bgColors = [
  [/^bg-(.+)$/, colorResolver$2("background-color", "bg")],
  [/^bg-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-bg-opacity": handler.bracket.percent(opacity2) })]
];
const borderColors = [
  [/^(?:border|b)-(.+)$/, colorResolver$2("border-color", "border")],
  [/^(?:border|b)-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-border-opacity": handler.bracket.percent(opacity2) })]
];
const ringColors = [
  [/^ring-(.+)$/, colorResolver$2("--un-ring-color", "ring")],
  [/^ring-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-ring-opacity": handler.bracket.percent(opacity2) })]
];
const ringOffsetColors = [
  [/^ring-offset-(.+)$/, colorResolver$2("--un-ring-offset-color", "ring-offset")],
  [/^ring-offset-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-ring-offset-opacity": handler.bracket.percent(opacity2) })]
];
const divideColors = [
  [/^divide-(.+)$/, colorResolver$2("border-color", "divide")],
  [/^divide-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-divide-opacity": handler.bracket.percent(opacity2) })]
];
const fillColors = [
  ["fill-none", { fill: "none" }],
  [/^fill-(.+)$/, colorResolver$2("fill", "fill")],
  [/^fill-op(?:acity)?-?(.+)$/m, ([, opacity2]) => ({ "--un-fill-opacity": handler.bracket.percent(opacity2) })]
];

const colorResolver$1 = (mode) => ([, body], { theme }) => {
  const data = parseColorUtil(body, theme);
  if (!data)
    return;
  const { opacity, color, rgba } = data;
  if (!color)
    return;
  let colorString = color;
  if (rgba) {
    const a = opacity ? opacity[0] === "[" ? handler.bracket.percent(opacity) : parseFloat(opacity) / 100 : rgba[3];
    if (a != null && !Number.isNaN(a)) {
      rgba[3] = typeof a === "string" && !a.includes("%") ? parseFloat(a) : a;
      colorString = `rgba(${rgba.join(",")})`;
    } else {
      colorString = `rgba(${rgba.slice(0, 3).join(",")}, var(--un-${mode}-opacity, 1))`;
    }
  }
  switch (mode) {
    case "from":
      return {
        "--un-gradient-from": colorString,
        "--un-gradient-stops": "var(--un-gradient-from), var(--un-gradient-to, rgba(255, 255, 255, 0))"
      };
    case "via":
      return {
        "--un-gradient-stops": `var(--un-gradient-from), ${colorString}, var(--un-gradient-to, rgba(255, 255, 255, 0))`
      };
    case "to":
      return {
        "--un-gradient-to": colorString
      };
  }
};
const bgAttachments = [
  ["bg-fixed", { "background-attachment": "fixed" }],
  ["bg-local", { "background-attachment": "local" }],
  ["bg-scroll", { "background-attachment": "scroll" }]
];
const bgBlendModes = [
  ["bg-blend-normal", { "background-blend-mode": "normal" }],
  ["bg-blend-multiply", { "background-blend-mode": "multiply" }],
  ["bg-blend-screen", { "background-blend-mode": "screen" }],
  ["bg-blend-overlay", { "background-blend-mode": "overlay" }],
  ["bg-blend-darken", { "background-blend-mode": "darken" }],
  ["bg-blend-lighten", { "background-blend-mode": "lighten" }],
  ["bg-blend-color-dodge", { "background-blend-mode": "color-dodge" }],
  ["bg-blend-color-burn", { "background-blend-mode": "color-burn" }],
  ["bg-blend-hard-light", { "background-blend-mode": "hard-light" }],
  ["bg-blend-soft-light", { "background-blend-mode": "soft-light" }],
  ["bg-blend-difference", { "background-blend-mode": "difference" }],
  ["bg-blend-exclusion", { "background-blend-mode": "exclusion" }],
  ["bg-blend-hue", { "background-blend-mode": "hue" }],
  ["bg-blend-saturation", { "background-blend-mode": "saturation" }],
  ["bg-blend-color", { "background-blend-mode": "color" }],
  ["bg-blend-luminosity", { "background-blend-mode": "luminosity" }]
];
const bgClips = [
  ["bg-clip-border", { "-webkit-background-clip": "border-box", "background-attachment": "border-box" }],
  ["bg-clip-content", { "-webkit-background-clip": "content-box", "background-attachment": "content-box" }],
  ["bg-clip-padding", { "-webkit-background-clip": "padding-box", "background-attachment": "padding-box" }],
  ["bg-clip-text", { "-webkit-background-clip": "text", "background-attachment": "text" }]
];
const bgGradients = [
  [/^from-(.+)$/, colorResolver$1("from")],
  [/^to-(.+)$/, colorResolver$1("to")],
  [/^via-(.+)$/, colorResolver$1("via")],
  [/^from-op(?:acity)?-?(.+)$/m, ([, opacity]) => ({ "--un-from-opacity": handler.bracket.percent(opacity) })],
  [/^to-op(?:acity)?-?(.+)$/m, ([, opacity]) => ({ "--un-to-opacity": handler.bracket.percent(opacity) })],
  [/^via-op(?:acity)?-?(.+)$/m, ([, opacity]) => ({ "--un-via-opacity": handler.bracket.percent(opacity) })]
];
const bgImages = [
  ["bg-none", { "background-image": "none" }],
  ["bg-gradient-to-t", {
    "background-image": "linear-gradient(to top, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-tr", {
    "background-image": "linear-gradient(to top right, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-r", {
    "background-image": "linear-gradient(to right, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-br", {
    "background-image": "linear-gradient(to bottom right, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-b", {
    "background-image": "linear-gradient(to bottom, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-bl", {
    "background-image": "linear-gradient(to bottom left, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-l", {
    "background-image": "linear-gradient(to left, var(--un-gradient-stops))"
  }],
  ["bg-gradient-to-tl", {
    "background-image": "linear-gradient(to top left, var(--un-gradient-stops))"
  }]
];
const bgOrigins = [
  ["bg-origin-border", { "background-origin": "border-box" }],
  ["bg-origin-padding", { "background-origin": "padding-box" }],
  ["bg-origin-content", { "background-origin": "content-box" }]
];
const bgPositions = [
  ["bg-bottom", { "background-position": "bottom" }],
  ["bg-center", { "background-position": "center" }],
  ["bg-left", { "background-position": "left" }],
  ["bg-left-bottom", { "background-position": "left bottom" }],
  ["bg-left-top", { "background-position": "left top" }],
  ["bg-right", { "background-position": "right" }],
  ["bg-right-bottom", { "background-position": "right bottom" }],
  ["bg-right-top", { "background-position": "right top" }],
  ["bg-top", { "background-position": "top" }]
];
const bgRepeats = [
  ["bg-repeat", { "background-repeat": "repeat" }],
  ["bg-no-repeat", { "background-repeat": "no-repeat" }],
  ["bg-repeat-x", { "background-position": "repeat-x" }],
  ["bg-repeat-y", { "background-position": "repeat-y" }],
  ["bg-repeat-round", { "background-position": "round" }],
  ["bg-repeat-space", { "background-position": "space" }]
];
const bgSizes = [
  ["bg-auto", { "background-size": "auto" }],
  ["bg-cover", { "background-repeat": "cover" }],
  ["bg-contain", { "background-position": "contain" }]
];

const outlineStyle = ["none", "auto", "dotted", "dashed", "solid", "double", "groove", "ridge", "inset", "outset", "inherit", "initial", "revert", "unset"];
const listStyleProps = ["none", "disc", "circle", "square", "decimal", "zero-decimal", "greek", "roman", "upper-roman", "alpha", "upper-alpha"];
const parseOutlineSize = (s) => {
  const propName = ["width", "offset"].find((item) => s.startsWith(item)) || "width";
  const size = handler.bracket.fraction.rem(s.replace(/^(offset\-|width\-)/, ""));
  if (size) {
    return {
      [`outline-${propName}`]: size
    };
  }
};
const outline = [
  ["outline-none", { "outline": "2px solid transparent", "outline-offset": "2px" }],
  ["outline", { "outline-style": "solid" }],
  [
    /^outline-(.+)$/,
    (match, config) => {
      const [, d] = match;
      if (d === "none") {
        return {
          "outline": "2px solid transparent",
          "outline-offset": "2px"
        };
      }
      if (outlineStyle.includes(d)) {
        return {
          "outline-style": d
        };
      }
      const sizeSheet = parseOutlineSize(d);
      if (sizeSheet)
        return sizeSheet;
      const colorSheet = colorResolver$2("outline-color", "outline-color")([
        match[0],
        match[1].replace(/^color-/, "")
      ], config);
      if (colorSheet)
        return colorSheet;
    }
  ]
];
const listStyle = [
  [new RegExp(`^list-((${listStyleProps.join("|")})(?:(-outside|-inside))?)$`), ([, value]) => {
    const style = value.split(/-outside|-inside/)[0];
    const position = /inside|outside/.exec(value) ?? [];
    if (position.length) {
      return {
        "list-style-position": `${position[0]}`,
        "list-style-type": `${style}`
      };
    }
    return {
      "list-style-type": `${style}`
    };
  }],
  [/^list-(inside|outside)$/, ([, value]) => {
    return {
      "list-style-position": value
    };
  }]
];
const boxDecorationBreaks = [
  ["decoration-slice", { "box-decoration-break": "slice" }],
  ["decoration-clone", { "box-decoration-break": "clone" }]
];
const caretOpacity = [
  [/^caret-op(?:acity)?-?(.+)$/, ([, d]) => ({ "--un-caret-opacity": handler.bracket.percent(d) })]
];
const caretColors = [
  [/^caret-(.+)$/, colorResolver$2("caret-color", "caret")]
];
const imageRenderings = [
  ["image-render-auto", { "image-rendering": "auto" }],
  ["image-render-edge", { "image-rendering": "crisp-edges" }],
  ["image-render-pixel", [
    ["-ms-interpolation-mode", "nearest-neighbor"],
    ["image-rendering", "-webkit-optimize-contrast"],
    ["image-rendering", "-moz-crisp-edges"],
    ["image-rendering", "-o-pixelated"],
    ["image-rendering", "pixelated"]
  ]]
];
const appearance = [
  ["appearance-none", {
    "appearance": "none",
    "-webkit-appearance": "none"
  }]
];
const placeholder = [
  [
    /^placeholder-opacity-(\d+)$/,
    ([, d]) => ({
      "placeholder-opacity": handler.bracket.percent(d)
    })
  ],
  [
    /^placeholder-(?!opacity)(.+)$/,
    (match, config) => {
      match[1] = match[1].replace(/^color-/, "");
      return colorResolver$2("placeholder-color", "placeholder-color")(match, config);
    }
  ]
];
const overflowValues$1 = [
  "none",
  "auto",
  "hidden",
  "visible",
  "scroll",
  "contain"
];
const overscrolls = [
  [/^overscroll-(.+)$/, ([, v]) => overflowValues$1.includes(v) ? { "overscroll-behavior": v } : void 0],
  [/^overscroll-([xy])-(.+)$/, ([, d, v]) => overflowValues$1.includes(v) ? { [`overscroll-behavior-${d}`]: v } : void 0]
];

const borderSizes = [
  [/^border$/, handlerBorder],
  [/^(?:border|b)(?:-([^-]+))?$/, handlerBorder],
  [/^(?:border|b)(?:-([^-]+))?(?:-([^-]+))?$/, handlerBorder]
];
const borderRadius = [
  [/^(?:border-)?(?:rounded|rd)$/, handlerRounded],
  [/^(?:border-)?(?:rounded|rd)(?:-([^-]+))?$/, handlerRounded],
  [/^(?:border-)?(?:rounded|rd)(?:-([^-]+))?(?:-([^-]+))?$/, handlerRounded]
];
const borderStyles = [
  ["border-solid", { "border-style": "solid" }],
  ["border-dashed", { "border-style": "dashed" }],
  ["border-dotted", { "border-style": "dotted" }],
  ["border-double", { "border-style": "double" }],
  ["border-none", { "border-style": "none" }]
];
const borders = [
  borderSizes,
  borderColors,
  borderStyles,
  borderRadius
].flat(1);
function handlerBorder([, a, b]) {
  const [d, s = "1"] = directionMap[a] ? [a, b] : ["", a];
  const v = handler.bracket.px(s);
  if (v != null) {
    return [
      ...directionMap[d].map((i) => [`border${i}-width`, v]),
      ["border-style", "solid"]
    ];
  }
}
function handlerRounded([, a, b], { theme }) {
  const [d, s = "DEFAULT"] = cornerMap[a] ? [a, b] : ["", a];
  const v = theme.borderRadius?.[s] || handler.bracket.fraction.rem(s);
  if (v != null)
    return cornerMap[d].map((i) => [`border${i}-radius`, v]);
}

const queryMatcher = /@media \(min-width: (.+)\)/;
const container = [
  [
    /^__container$/,
    (m, { variantHandlers }) => {
      let width = "100%";
      for (const v of variantHandlers) {
        const query = toArray(v.parent || [])[0];
        if (typeof query === "string") {
          const match = query.match(queryMatcher)?.[1];
          if (match)
            width = match;
        }
      }
      return { "max-width": width };
    },
    { internal: true }
  ]
];
const containerShortcuts = [
  [/^(?:(\w+)[:-])?container$/, ([, bp], { theme }) => {
    let points = Object.keys(theme.breakpoints || {});
    if (bp) {
      if (!points.includes(bp))
        return;
      points = points.slice(points.indexOf(bp));
    }
    const shortcuts = points.map((p) => `${p}:__container`);
    if (!bp)
      shortcuts.unshift("__container");
    return shortcuts;
  }]
];

const transitionBasicProps = [
  "color",
  "border-color",
  "background-color",
  "flex-grow",
  "flex",
  "flex-shrink",
  "caret-color",
  "font",
  "gap",
  "opacity",
  "visibility",
  "z-index",
  "font-weight",
  "zoom",
  "text-shadow",
  "transform",
  "box-shadow"
];
const transitionPositionProps = [
  "backround-position",
  "left",
  "right",
  "top",
  "bottom",
  "object-position"
];
const transitionSizeProps = [
  "max-height",
  "min-height",
  "max-width",
  "min-width",
  "height",
  "width",
  "border-width",
  "margin",
  "padding",
  "outline-width",
  "outline-offset",
  "font-size",
  "line-height",
  "text-indent",
  "vertical-align",
  "border-spacing",
  "letter-spacing",
  "word-spacing"
];
const transitionSwitchProps = ["all", "none"];
const transitionEnhanceProps = ["stroke", "filter", "backdrop-filter", "fill", "mask", "mask-size", "mask-border", "clip-path", "clip"];
const transitionProps = [
  ...transitionBasicProps,
  ...transitionPositionProps,
  ...transitionSizeProps,
  ...transitionEnhanceProps
];
const transitionPropsStr = transitionProps.join(", ");
const validateProperty = (prop) => {
  if (prop && ![...transitionProps, ...transitionSwitchProps].includes(prop))
    return;
  return prop || transitionPropsStr;
};
const transitions = [
  [/^transition(?:-([a-z-]+))?(?:-(\d+))?$/, ([, prop, duration = "150"]) => {
    const transitionProperty = validateProperty(prop);
    if (!transitionProperty)
      return;
    return {
      "transition-property": transitionProperty,
      "transition-timing-function": "cubic-bezier(0.4, 0, 0.2, 1)",
      "transition-duration": `${duration}ms`
    };
  }],
  [/^duration-(\d+)$/, ([, duration = "150"]) => {
    return {
      "transition-duration": `${duration}ms`
    };
  }],
  ["ease", { "transition-timing-function": "cubic-bezier(0.4, 0, 0.2, 1)" }],
  ["ease-in", { "transition-timing-function": "cubic-bezier(0.4, 0, 1, 1)" }],
  ["ease-out", { "transition-timing-function": "cubic-bezier(0, 0, 0.2, 1)" }],
  ["ease-in-out", { "transition-timing-function": "cubic-bezier(0.4, 0, 0.2, 1)" }],
  [/^transition-delay-(\d+)$/, ([, v]) => ({ "transition-delay": `${v}ms` })],
  [/^transition-duration-(\d+)$/, ([, v]) => ({ "transition-duration": `${v}ms` })],
  [/^(?:transition-)?property-([a-z-]+)$/, ([, v]) => {
    const transitionProperty = validateProperty(v);
    if (transitionProperty)
      return { "transition-property": transitionProperty };
  }]
];

const varEmpty$1 = "var(--un-empty,/*!*/ /*!*/)";
const filterContnet = "var(--un-blur) var(--un-brightness) var(--un-contrast) var(--un-grayscale) var(--un-hue-rotate) var(--un-invert) var(--un-saturate) var(--un-sepia) var(--un-drop-shadow)";
const filterBase = {
  "--un-blur": varEmpty$1,
  "--un-brightness": varEmpty$1,
  "--un-contrast": varEmpty$1,
  "--un-grayscale": varEmpty$1,
  "--un-hue-rotate": varEmpty$1,
  "--un-invert": varEmpty$1,
  "--un-saturate": varEmpty$1,
  "--un-sepia": varEmpty$1,
  "--un-drop-shadow": varEmpty$1,
  "filter": filterContnet,
  [CONTROL_BYPASS_PSEUDO]: ""
};
const backdropFilterContent = "var(--un-backdrop-blur) var(--un-backdrop-brightness) var(--un-backdrop-contrast) var(--un-backdrop-grayscale) var(--un-backdrop-hue-rotate) var(--un-backdrop-invert) var(--un-backdrop-saturate) var(--un-backdrop-sepia)";
const backdropFilterBase = {
  "--un-backdrop-blur": varEmpty$1,
  "--un-backdrop-brightness": varEmpty$1,
  "--un-backdrop-contrast": varEmpty$1,
  "--un-backdrop-grayscale": varEmpty$1,
  "--un-backdrop-hue-rotate": varEmpty$1,
  "--un-backdrop-invert": varEmpty$1,
  "--un-backdrop-saturate": varEmpty$1,
  "--un-backdrop-sepia": varEmpty$1,
  "-webkit-backdrop-filter": backdropFilterContent,
  "backdrop-filter": backdropFilterContent,
  [CONTROL_BYPASS_PSEUDO]: ""
};
const percentWithDefault = (defaultValue = "1") => (str) => {
  const v = str ? handler.bracket.percent(str) : defaultValue;
  return v && parseFloat(v) <= 1 ? v : void 0;
};
const toFilter = (varName, resolver) => ([, b, s], { theme }) => {
  const value = resolver(s, theme);
  if (value) {
    return [
      b ? backdropFilterBase : filterBase,
      { [`--un-${b || ""}${varName}`]: `${varName}(${value})` }
    ];
  }
};
const filters = [
  ["filter", filterBase],
  ["filter-none", { filter: "none" }],
  ["backdrop-filter", backdropFilterBase],
  ["backdrop-filter-none", {
    "-webkit-backdrop-filter": "none",
    "backdrop-filter": "none"
  }],
  [/^(backdrop-)?blur(?:-(.+))?$/, toFilter("blur", (s, theme) => theme.blur?.[s || "DEFAULT"] || handler.bracket.px(s))],
  [/^(backdrop-)?brightness-(\d+)$/, toFilter("brightness", (s) => handler.bracket.percent(s))],
  [/^(backdrop-)?contrast-(\d+)$/, toFilter("contrast", (s) => handler.bracket.percent(s))],
  [/^()?drop-shadow(?:-(.+))?$/, toFilter("drop-shadow", (s, theme) => {
    const v = handler.bracket(s) || theme.dropShadow?.[s || "DEFAULT"];
    if (v)
      return toArray(v).map((v2) => `drop-shadow(${v2})`).join(" ");
  })],
  [/^(backdrop-)?grayscale(?:-(\d+))?$/, toFilter("grayscale", percentWithDefault())],
  [/^(backdrop-)?hue-rotate-(\d+)$/, toFilter("hue-rotate", (s) => `${handler.bracket.number(s)}deg`)],
  [/^(backdrop-)?invert(?:-(\d+))?$/, toFilter("invert", percentWithDefault())],
  [/^(backdrop-)?saturate(?:-(\d+))?$/, toFilter("saturate", percentWithDefault("0"))],
  [/^(backdrop-)?sepia(?:-(\d+))?$/, toFilter("sepia", percentWithDefault())]
];

const flex = [
  ["flex-col", { "flex-direction": "column" }],
  ["flex-col-reverse", { "flex-direction": "column-reverse" }],
  ["flex-row", { "flex-direction": "row" }],
  ["flex-row-reverse", { "flex-direction": "row-reverse" }],
  ["flex-wrap", { "flex-wrap": "wrap" }],
  ["flex-wrap-reverse", { "flex-wrap": "wrap-reverse" }],
  ["flex-nowrap", { "flex-wrap": "nowrap" }],
  ["flex-1", { flex: "1 1 0%" }],
  ["flex-auto", { flex: "1 1 auto" }],
  ["flex-initial", { flex: "0 1 auto" }],
  ["flex-none", { flex: "none" }],
  [/^flex-\[(.+)\]$/, ([, d]) => ({ flex: d })],
  ["flex-grow", { "flex-grow": 1 }],
  ["flex-grow-0", { "flex-grow": 0 }],
  ["flex-shrink", { "flex-shrink": 1 }],
  ["flex-shrink-0", { "flex-shrink": 0 }],
  ["flex", { display: "flex" }],
  ["inline-flex", { display: "inline-flex" }],
  ["flex-inline", { display: "inline-flex" }]
];

const varEmpty = "var(--un-empty,/*!*/ /*!*/)";
const displays = [
  ["inline", { display: "inline" }],
  ["block", { display: "block" }],
  ["inline-block", { display: "inline-block" }],
  ["contents", { display: "contents" }],
  ["flow-root", { display: "flow-root" }],
  ["list-item", { display: "list-item" }],
  ["hidden", { display: "none" }]
];
const appearances = [
  ["visible", { visibility: "visible" }],
  ["invisible", { visibility: "hidden" }],
  ["backface-visible", { "backface-visibility": "visible" }],
  ["backface-hidden", { "backface-visibility": "hidden" }]
];
const cursors = [
  [/^cursor-(.+)$/, ([, c]) => ({ cursor: c })]
];
const pointerEvents = [
  ["pointer-events-none", { "pointer-events": "none" }],
  ["pointer-events-auto", { "pointer-events": "auto" }]
];
const resizes = [
  ["resize-none", { resize: "none" }],
  ["resize-x", { resize: "horizontal" }],
  ["resize-y", { resize: "vertical" }],
  ["resize", { resize: "both" }]
];
const userSelects = [
  [/^select-(none|text|all|auto)$/, ([, v]) => ({ "user-select": v })]
];
const whitespaces = [
  [/^(?:whitespace|ws)-(normal|nowrap|pre|pre-line|pre-wrap)$/, ([, v]) => ({ "white-space": v })]
];
const contents = [
  ["content-empty", { content: '""' }]
];
const breaks = [
  ["break-normal", { "overflow-wrap": "normal", "word-break": "normal" }],
  ["break-word", { "overflow-wrap": "break-word" }],
  ["break-all", { "word-break": "break-all" }]
];
const textOverflows = [
  ["truncate", { "overflow": "hidden", "text-overflow": "ellipsis", "white-space": "nowrap" }],
  ["text-ellipsis", { "text-overflow": "ellipsis" }],
  ["text-clip", { "text-overflow": "clip" }]
];
const textTransforms = [
  ["case-upper", { "text-transform": "uppercase" }],
  ["case-lower", { "text-transform": "lowercase" }],
  ["case-capital", { "text-transform": "capitalize" }],
  ["case-normal", { "text-transform": "none" }],
  ["uppercase", { "text-transform": "uppercase" }],
  ["lowercase", { "text-transform": "lowercase" }],
  ["capitalize", { "text-transform": "capitalize" }],
  ["normal-case", { "text-transform": "none" }]
];
const textDecorations = [
  ["underline", { "text-decoration": "underline" }],
  ["line-through", { "text-decoration": "line-through" }],
  ["no-underline", { "text-decoration": "none" }]
];
const textDecorationStyles = [
  ["underline-solid", { "text-decoration-style": "solid" }],
  ["underline-double", { "text-decoration-style": "double" }],
  ["underline-dotted", { "text-decoration-style": "dotted" }],
  ["underline-dashed", { "text-decoration-style": "dashed" }]
];
const fontStyles = [
  ["italic", { "font-style": "italic" }],
  ["not-italic", { "font-style": "normal" }]
];
const fontSmoothings = [
  ["antialiased", {
    "-webkit-font-smoothing": "antialiased",
    "-moz-osx-font-smoothing": "grayscale",
    "font-smoothing": "grayscale"
  }],
  ["subpixel-antialiased", {
    "-webkit-font-smoothing": "auto",
    "-moz-osx-font-smoothing": "auto",
    "font-smoothing": "auto"
  }]
];
const hyphens = [
  ["hyphens-none", {
    "-webkit-hyphens": "none",
    "-ms-hyphens": "none",
    "hyphens": "none"
  }],
  ["hyphens-manual", {
    "-webkit-hyphens": "manual",
    "-ms-hyphens": "manual",
    "hyphens": "manual"
  }],
  ["hyphens-auto", {
    "-webkit-hyphens": "auto",
    "-ms-hyphens": "auto",
    "hyphens": "auto"
  }]
];
const writingModes = [
  ["write-normal", { "writing-mode": "horizontal-tb" }],
  ["write-vertical-right", { "writing-mode": "vertical-rl" }],
  ["write-vertical-left", { "writing-mode": "vertical-lr" }]
];
const writingOrientations = [
  ["write-orient-mixed", { "text-orientation": "mixed" }],
  ["write-orient-upright", { "text-orientation": "upright" }],
  ["write-orient-sideways", { "text-orientation": "sideways" }]
];
const screenReadersAccess = [
  [
    "sr-only",
    {
      "position": "absolute",
      "width": "1px",
      "height": "1px",
      "padding": "0",
      "margin": "-1px",
      "overflow": "hidden",
      "clip": "rect(0,0,0,0)",
      "white-space": "nowrap",
      "border-width": 0
    }
  ],
  [
    "not-sr-only",
    {
      "position": "static",
      "width": "auto",
      "height": "auto",
      "padding": "0",
      "margin": "0",
      "overflow": "visible",
      "clip": "auto",
      "white-space": "normal"
    }
  ]
];
const isolations = [
  ["isolate", { isolation: "isolate" }],
  ["isolate-auto", { isolation: "auto" }]
];

const fontsFamilies = [
  [/^font-(\w+)$/, ([, d], { theme }) => {
    const font = theme.fontFamily?.[d];
    if (font) {
      return {
        "font-family": font
      };
    }
  }]
];
const weightMap = {
  thin: "100",
  extralight: "200",
  light: "300",
  normal: "400",
  medium: "500",
  semibold: "600",
  bold: "700",
  extrabold: "800",
  black: "900"
};
const fontSizes = [
  [/^text-([^-]+)$/, ([, s = "base"], { theme }) => {
    const result = toArray(theme.fontSize?.[s] || handler.bracket.rem(s));
    if (result?.[0]) {
      const [size, height = "1"] = result;
      return {
        "font-size": size,
        "line-height": height
      };
    }
  }]
];
const fontWeights = [
  [/^(?:font|fw)-?([^-]+)$/, ([, s]) => {
    const v = weightMap[s] || handler.number(s);
    if (v)
      return { "font-weight": v };
  }]
];
const leadings = [
  [/^(?:leading|lh)-([^-]+)$/, ([, s], { theme }) => {
    const v = theme.lineHeight?.[s] || handler.bracket.rem(s);
    if (v !== null)
      return { "line-height": v };
  }]
];
const trackings = [
  [/^tracking-([^-]+)$/, ([, s], { theme }) => {
    const v = theme.letterSpacing?.[s] || handler.bracket.rem(s);
    if (v !== null)
      return { "letter-spacing": v };
  }]
];
const wordSpacings = [
  [/^word-spacing-([^-]+)$/, ([, s], { theme }) => {
    const v = theme.wordSpacing?.[s] || handler.bracket.rem(s);
    if (v !== null)
      return { "word-spacing": v };
  }]
];
const tabSizes = [
  [/^tab-?([^-]*)$/, ([, s]) => {
    s = s || "4";
    const v = handler.bracket.global.number(s);
    if (v !== null) {
      return {
        "-moz-tab-size": v,
        "-o-tab-size": v,
        "tab-size": v
      };
    }
  }]
];
const textDecorationLengths = [
  [/^underline-([^-]+)$/, ([, s]) => {
    const v = s === "auto" ? s : handler.bracket.px(s);
    if (v != null)
      return { "text-decoration-thickness": v };
  }]
];
const textDecorationOffsets = [
  [/^underline-offset-([^-]+)$/, ([, s]) => {
    const v = s === "auto" ? s : handler.bracket.px(s);
    if (v != null)
      return { "text-underline-offset": v };
  }]
];
const textIndents = [
  [/^indent(?:-(.+))?$/, ([, s], { theme }) => {
    const v = theme.textIndent?.[s || "DEFAULT"] || handler.bracket.cssvar.fraction.rem(s);
    if (v != null)
      return { "text-indent": v };
  }]
];
const textStrokeWidths = [
  [/^text-stroke(?:-(.+))?$/, ([, s], { theme }) => {
    const v = theme.textStrokeWidth?.[s || "DEFAULT"] || handler.bracket.cssvar.px(s);
    if (v != null)
      return { "-webkit-text-stroke-width": v };
  }]
];
const textShadows = [
  [/^text-shadow(?:-(.+))?$/, ([, s], { theme }) => {
    const v = theme.textShadow?.[s || "DEFAULT"] || handler.bracket.cssvar(s);
    if (v != null)
      return { "text-shadow": v };
  }]
];
const fontVariantNumericBase = {
  "--un-ordinal": varEmpty,
  "--un-slashed-zero": varEmpty,
  "--un-numeric-figure": varEmpty,
  "--un-numeric-spacing": varEmpty,
  "--un-numeric-fraction": varEmpty,
  "font-variant-numeric": "var(--un-ordinal) var(--un-slashed-zero) var(--un-numeric-figure) var(--un-numeric-spacing) var(--un-numeric-fraction)",
  [CONTROL_BYPASS_PSEUDO]: ""
};
const fontVariantNumeric = [
  [/^ordinal$/, () => [fontVariantNumericBase, { "--un-ordinal": "ordinal" }]],
  [/^slashed-zero$/, () => [fontVariantNumericBase, { "--un-slashed-zero": "slashed-zero" }]],
  [/^lining-nums$/, () => [fontVariantNumericBase, { "--un-numeric-figure": "lining-nums" }]],
  [/^oldstyle-nums$/, () => [fontVariantNumericBase, { "--un-numeric-figure": "oldstyle-nums" }]],
  [/^proportional-nums$/, () => [fontVariantNumericBase, { "--un-numeric-spacing": "proportional-nums" }]],
  [/^tabular-nums$/, () => [fontVariantNumericBase, { "--un-numeric-spacing": "tabular-nums" }]],
  [/^diagonal-fractions$/, () => [fontVariantNumericBase, { "--un-numeric-fraction": "diagonal-fractions" }]],
  [/^stacked-fractions$/, () => [fontVariantNumericBase, { "--un-numeric-fraction": "stacked-fractions" }]],
  ["normal-nums", { "font-variant-numeric": "normal" }]
];
const fonts = [
  fontsFamilies,
  fontSizes,
  fontWeights
].flat(1);

const gaps = [
  [/^(?:flex-|grid-)?gap-([^-]+)$/, ([, s]) => {
    const v = handler.bracket.rem(s);
    if (v != null) {
      return {
        "grid-gap": v,
        "gap": v
      };
    }
  }],
  [/^(?:flex-|grid-)?gap-x-([^-]+)$/, ([, s]) => {
    const v = handler.bracket.rem(s);
    if (v != null) {
      return {
        "grid-column-gap": v,
        "column-gap": v
      };
    }
  }],
  [/^(?:flex-|grid-)?gap-y-([^-]+)$/, ([, s]) => {
    const v = handler.bracket.rem(s);
    if (v != null) {
      return {
        "grid-row-gap": v,
        "row-gap": v
      };
    }
  }]
];

const calSize = (s, theme) => toArray(theme.fontSize?.[s] || handler.bracket.rem(s))[0];
const autoDirection = (selector, theme) => {
  if (selector === "min")
    return "min-content";
  else if (selector === "max")
    return "max-content";
  else if (selector === "fr")
    return "minmax(0,1fr)";
  return calSize(selector, theme);
};
const grids = [
  ["grid", { display: "grid" }],
  ["inline-grid", { display: "inline-grid" }],
  [/^grid-cols-minmax-([\w.-]+)$/, ([, d]) => ({ "grid-template-columns": `repeat(auto-fill, minmax(${d}, 1fr))` })],
  [/^grid-rows-minmax-([\w.-]+)$/, ([, d]) => ({ "grid-template-rows": `repeat(auto-fill, minmax(${d}, 1fr))` })],
  [/^grid-cols-(\d+)$/, ([, d]) => ({ "grid-template-columns": `repeat(${d},minmax(0,1fr))` })],
  [/^grid-rows-(\d+)$/, ([, d]) => ({ "grid-template-rows": `repeat(${d},minmax(0,1fr))` })],
  [/^grid-cols-\[(.+)\]$/, ([, v]) => ({ "grid-template-columns": v.replace(/,/g, " ") })],
  [/^grid-rows-\[(.+)\]$/, ([, v]) => ({ "grid-template-rows": v.replace(/,/g, " ") })],
  [/^(?:grid-)?(row|col)-(.+)$/, ([, d, v]) => {
    const key = d === "row" ? "grid-row" : "grid-column";
    let raw = handler.bracket(v);
    if (raw)
      return { [key]: raw };
    const parts = v.split("-");
    if (parts.length === 1 && parts[0] === "auto")
      return { [key]: parts[0] };
    if (parts[0] === "span") {
      if (parts[1] === "full")
        return { [key]: "1/-1" };
      raw = handler.number.bracket(parts[1])?.toString().replace(/_/g, " ");
      if (raw)
        return { [key]: `span ${raw}/span ${raw}` };
    }
  }],
  [/^(?:grid-)?auto-flow-([\w.-]+)$/, ([, v]) => ({ "grid-auto-flow": `${v.replace("col", "column").split("-").join(" ")}` })],
  [/^(?:grid-)?row-start-([\w.-]+)$/, ([, v]) => ({ "grid-row-start": `${v}` })],
  [/^(?:grid-)?row-end-([\w.-]+)$/, ([, v]) => ({ "grid-row-end": `${v}` })],
  [/^(?:grid-)?col-start-([\w.-]+)$/, ([, v]) => ({ "grid-column-start": `${v}` })],
  [/^(?:grid-)?col-end-([\w.]+)$/, ([, v]) => ({ "grid-column-end": `${v}` })],
  [/^(?:grid-)?auto-rows-([\w.-]+)$/, ([, v], { theme }) => ({ "grid-auto-rows": `${autoDirection(v, theme)}` })],
  [/^(?:grid-)?auto-cols-([\w.-]+)$/, ([, v], { theme }) => ({ "grid-auto-columns": `${autoDirection(v, theme)}` })]
];

const overflowValues = [
  "auto",
  "hidden",
  "visible",
  "scroll"
];
const overflows = [
  [/^(?:overflow|of)-(.+)$/, ([, v]) => overflowValues.includes(v) ? { overflow: v } : void 0],
  [/^(?:overflow|of)-([xy])-(.+)$/, ([, d, v]) => overflowValues.includes(v) ? { [`overflow-${d}`]: v } : void 0]
];

const positions = [
  ["relative", { position: "relative" }],
  ["absolute", { position: "absolute" }],
  ["fixed", { position: "fixed" }],
  ["sticky", { position: "sticky" }],
  ["static", { position: "static" }]
];
const justifies = [
  ["justify-start", { "justify-content": "flex-start" }],
  ["justify-end", { "justify-content": "flex-end" }],
  ["justify-center", { "justify-content": "center" }],
  ["justify-between", { "justify-content": "space-between" }],
  ["justify-around", { "justify-content": "space-around" }],
  ["justify-evenly", { "justify-content": "space-evenly" }]
];
const orders = [
  [/^order-(.+)$/, ([, v]) => ({ order: { first: "-9999", last: "9999", none: "0" }[v] || handler.bracket.number(v) })]
];
const basicSet = ["auto", "start", "end", "center", "stretch"];
const justifyItems = basicSet.map((i) => [`justify-items-${i}`, { "justify-items": i }]);
const justifySelfs = basicSet.map((i) => [`justify-self-${i}`, { "justify-self": i }]);
const alignContents = [
  ["content-start", { "align-content": "flex-start" }],
  ["content-end", { "align-content": "flex-end" }],
  ["content-center", { "align-content": "center" }],
  ["content-between", { "align-content": "space-between" }],
  ["content-around", { "align-content": "space-around" }],
  ["content-evenly", { "align-content": "space-evenly" }]
];
const alignItems = [
  ["items-start", { "align-items": "flex-start" }],
  ["items-end", { "align-items": "flex-end" }],
  ["items-center", { "align-items": "center" }],
  ["items-baseline", { "align-items": "baseline" }],
  ["items-stretch", { "align-items": "stretch" }]
];
const alignSelfs = [
  ["self-auto", { "align-self": "auto" }],
  ["self-start", { "align-self": "flex-start" }],
  ["self-end", { "align-self": "flex-end" }],
  ["self-center", { "align-self": "center" }],
  ["self-stretch", { "align-items": "stretch" }]
];
const placeContents = [
  ["place-content-start", { "place-content": "start" }],
  ["place-content-end", { "place-content": "end" }],
  ["place-content-center", { "place-content": "center" }],
  ["place-content-between", { "place-content": "space-between" }],
  ["place-content-around", { "place-content": "space-around" }],
  ["place-content-evenly", { "place-content": "space-evenly" }],
  ["place-content-stretch", { "place-content": "stretch" }]
];
const placeItems = basicSet.map((i) => [`place-items-${i}`, { "place-items": i }]);
const placeSelfs = basicSet.map((i) => [`place-self-${i}`, { "place-self": i }]);
function handleInsetValue(v) {
  return { auto: "auto", full: "100%" }[v] ?? handler.bracket.fraction.cssvar.rem(v);
}
const insets = [
  [/^(top|left|right|bottom|inset)-(.+)$/, ([, d, v]) => ({ [d]: handleInsetValue(v) })],
  [/^inset-([xy])-(.+)$/, ([, d, v]) => {
    const r = handleInsetValue(v);
    if (r != null && d in directionMap)
      return directionMap[d].map((i) => [i.slice(1), r]);
  }]
];
const floats = [
  [/^float-(left|right|none)$/, ([, value]) => ({ float: value })],
  [/^clear-(left|right|both|none)$/, ([, value]) => ({ clear: value })]
];
const zIndexes = [
  ["z-auto", { "z-index": "auto" }],
  [/^z-([^-]+)$/, ([, v]) => ({ "z-index": handler.number(v) })]
];
const objectPositions = [
  ["object-cover", { "object-fit": "cover" }],
  ["object-contain", { "object-fit": "contain" }],
  ["object-fill", { "object-fit": "fill" }],
  ["object-scale-down", { "object-fit": "scale-down" }],
  ["object-none", { "object-fit": "none" }],
  ["object-center", { "object-position": "center" }],
  ["object-bottom", { "object-position": "bottom" }],
  ["object-top", { "object-position": "top" }],
  ["object-right", { "object-position": "right" }],
  ["object-left", { "object-position": "left" }],
  ["object-lb", { "object-position": "left bottom" }],
  ["object-lt", { "object-position": "left top" }],
  ["object-rb", { "object-position": "right bottom" }],
  ["object-rt", { "object-position": "right top" }],
  ["object-cb", { "object-position": "center bottom" }],
  ["object-ct", { "object-position": "center top" }]
];
const boxSizing = [
  [
    /^box-(border|content)$/,
    ([, value]) => ({
      "box-sizing": `${value}-box`
    })
  ]
];

const rings = [
  [/^ring-?(.*)$/, ([, d]) => {
    const value = handler.px(d || "1");
    if (value) {
      return {
        "--un-ring-inset": "var(--un-empty, )",
        "--un-ring-offset-width": "0px",
        "--un-ring-offset-color": "#fff",
        "--un-ring-color": "rgba(59, 130, 246, .5)",
        "--un-ring-offset-shadow": "var(--un-ring-inset) 0 0 0 var(--un-ring-offset-width) var(--un-ring-offset-color)",
        "--un-ring-shadow": `var(--un-ring-inset) 0 0 0 calc(${value} + var(--un-ring-offset-width)) var(--un-ring-color)`,
        "-webkit-box-shadow": "var(--un-ring-offset-shadow), var(--un-ring-shadow), var(--un-shadow, 0 0 #0000);",
        "box-shadow": "var(--un-ring-offset-shadow), var(--un-ring-shadow), var(--un-shadow, 0 0 #0000);"
      };
    }
  }],
  [/^ring-offset-?(.*)$/, ([, d]) => {
    const value = handler.px(d || "1");
    if (value) {
      return {
        "--un-ring-offset-width": value
      };
    }
  }],
  ["ring-inset", { "--un-ring-inset": "inset" }],
  ...ringColors,
  ...ringOffsetColors
];

const colorResolver = (body, theme) => {
  const data = parseColorUtil(body, theme);
  if (!data)
    return;
  const { color, rgba } = data;
  if (!color)
    return;
  if (rgba) {
    return {
      "--un-shadow-color": `${rgba.slice(0, 3).join(",")}`
    };
  } else {
    return {
      "--un-shadow-color": color
    };
  }
};
const mixBlendModes = [
  ["mix-blend-normal", { "mix-blend-mode": "normal" }],
  ["mix-blend-multiply", { "mix-blend-mode": "multiply" }],
  ["mix-blend-screen", { "mix-blend-mode": "screen" }],
  ["mix-blend-overlay", { "mix-blend-mode": "overlay" }],
  ["mix-blend-darken", { "mix-blend-mode": "darken" }],
  ["mix-blend-lighten", { "mix-blend-mode": "lighten" }],
  ["mix-blend-color-dodge", { "mix-blend-mode": "color-dodge" }],
  ["mix-blend-color-burn", { "mix-blend-mode": "color-burn" }],
  ["mix-blend-color-light", { "mix-blend-mode": "color-light" }],
  ["mix-blend-soft-light", { "mix-blend-mode": "soft-light" }],
  ["mix-blend-difference", { "mix-blend-mode": "difference" }],
  ["mix-blend-exclusion", { "mix-blend-mode": "exclusion" }],
  ["mix-blend-hue", { "mix-blend-mode": "hue" }],
  ["mix-blend-saturation", { "mix-blend-mode": "saturation" }],
  ["mix-blend-color", { "mix-blend-mode": "color" }],
  ["mix-blend-luminosity", { "mix-blend-mode": "luminosity" }]
];
const boxShadows = [
  [/^shadow-?(.*)$/, ([, d], { theme }) => {
    const value = theme?.boxShadow?.[d || "DEFAULT"];
    if (value) {
      return {
        "--un-shadow-color": "0,0,0",
        "--un-shadow": value,
        "box-shadow": "var(--un-ring-offset-shadow, 0 0 #0000), var(--un-ring-shadow, 0 0 #0000), var(--un-shadow)"
      };
    }
    const color = colorResolver(d, theme);
    if (color)
      return color;
  }]
];

function getPropName(minmax, hw) {
  return `${minmax ? `${minmax}-` : ""}${hw === "h" ? "height" : "width"}`;
}
function getThemeValue(minmax, hw, theme, prop) {
  let str = `${hw === "h" ? "height" : "width"}`;
  if (minmax)
    str = `${minmax}${capitalize(str)}`;
  return theme[str]?.[prop];
}
const sizes = [
  [/^(?:(min|max)-)?(w|h)-(.+)$/, ([, m, w, s], { theme }) => {
    const v = getThemeValue(m, w, theme, s) || handler.bracket.cssvar.fraction.rem(s);
    if (v != null)
      return { [getPropName(m, w)]: v };
  }],
  [/^(?:(min|max)-)?(w)-screen-(.+)$/, ([, m, w, s], { theme }) => {
    const v = theme.breakpoints?.[s];
    if (v != null)
      return { [getPropName(m, w)]: v };
  }]
];
const aspectRatio = [
  ["aspect-ratio-auto", { "aspect-ratio": "auto" }],
  [/^aspect-ratio-(.+)$/, ([, d]) => {
    const v = (/^\d+\/\d+$/.test(d) ? d : null) || handler.bracket.cssvar.number(d);
    if (v != null)
      return { "aspect-ratio": v };
  }]
];

const directionSize = (prefix) => ([_, direction, size]) => {
  const v = handler.bracket.rem.fraction.cssvar(size);
  if (v)
    return directionMap[direction].map((i) => [prefix + i, v]);
};
const paddings = [
  [/^pa?()-?(-?.+)$/, directionSize("padding")],
  [/^p-?([xy])-?(-?.+)$/, directionSize("padding")],
  [/^p-?([rltbse])-?(-?.+)$/, directionSize("padding")]
];
const margins = [
  [/^ma?()-?(-?.+)$/, directionSize("margin")],
  [/^m-?([xy])-?(-?.+)$/, directionSize("margin")],
  [/^m-?([rltbse])-?(-?.+)$/, directionSize("margin")],
  [/^space-?([xy])-?(-?.+)$/, (match) => {
    const [, direction, size] = match;
    if (size === "reverse")
      return { [`--un-space-${direction}-reverse`]: 1 };
    const results = directionSize("margin")(match)?.map((item) => {
      const value = item[0].endsWith("right") || item[0].endsWith("bottom") ? `calc(${item[1]} * var(--un-space-${direction}-reverse))` : `calc(${item[1]} * calc(1 - var(--un-space-${direction}-reverse)))`;
      return [item[0], value];
    });
    if (results) {
      return [
        [`--un-space-${direction}-reverse`, 0],
        ...results
      ];
    }
  }]
];

const tables = [
  ["border-collapse", { "border-collapse": "collapse" }],
  ["border-separate", { "border-collapse": "separate" }],
  ["caption-top", { "caption-side": "top" }],
  ["caption-bottom", { "caption-side": "bottom" }],
  ["inline-table", { display: "inline-table" }],
  ["table", { display: "table" }],
  ["table-auto", { "table-layout": "auto" }],
  ["table-empty-cells-visible", { "empty-cells": "show" }],
  ["table-empty-cells-hidden", { "empty-cells": "hide" }],
  ["table-fixed", { "table-layout": "fixed" }],
  ["table-caption", { display: "table-caption" }],
  ["table-cell", { display: "table-cell" }],
  ["table-column", { display: "table-column" }],
  ["table-column-group", { display: "table-column-group" }],
  ["table-footer-group", { display: "table-footer-group" }],
  ["table-header-group", { display: "table-header-group" }],
  ["table-row", { display: "table-row" }],
  ["table-row-group", { display: "table-row-group" }]
];

const transformBase = {
  "--un-rotate": 0,
  "--un-scale-x": 1,
  "--un-scale-y": 1,
  "--un-scale-z": 1,
  "--un-skew-x": 0,
  "--un-skew-y": 0,
  "--un-translate-x": 0,
  "--un-translate-y": 0,
  "--un-translate-z": 0,
  "transform": "rotate(var(--un-rotate)) scaleX(var(--un-scale-x)) scaleY(var(--un-scale-y)) scaleZ(var(--un-scale-z)) skewX(var(--un-skew-x)) skewY(var(--un-skew-y)) translateX(var(--un-translate-x)) translateY(var(--un-translate-y)) translateZ(var(--un-translate-z))",
  [CONTROL_BYPASS_PSEUDO]: ""
};
const transforms = [
  ["transform", transformBase],
  [/^preserve-(3d|flat)$/, ([, value]) => ({ "transform-style": value === "3d" ? `preserve-${value}` : value })],
  [/^translate()-([^-]+)$/, handleTranslate],
  [/^translate-([xyz])-([^-]+)$/, handleTranslate],
  [/^scale()-([^-]+)$/, handleScale],
  [/^scale-([xyz])-([^-]+)$/, handleScale],
  [/^rotate-([^-]+)(?:deg)?$/, handleRotate],
  ["origin-center", { "transform-origin": "center" }],
  ["origin-top", { "transform-origin": "top" }],
  ["origin-top-right", { "transform-origin": "top right" }],
  ["origin-right", { "transform-origin": "right" }],
  ["origin-bottom-right", { "transform-origin": "bottom right" }],
  ["origin-bottom", { "transform-origin": "bottom" }],
  ["origin-bottom-left", { "transform-origin": "bottom left" }],
  ["origin-left", { "transform-origin": "left" }],
  ["origin-top-left", { "transform-origin": "top left" }]
];
function handleTranslate([, d, b]) {
  const v = handler.bracket.fraction.rem(b);
  if (v != null) {
    return [
      transformBase,
      [
        ...xyzMap[d].map((i) => [`--un-translate${i}`, v])
      ]
    ];
  }
}
function handleScale([, d, b]) {
  const v = handler.bracket.fraction.percent(b);
  if (v != null) {
    return [
      transformBase,
      [
        ...xyzMap[d].map((i) => [`--un-scale${i}`, v])
      ]
    ];
  }
}
function handleRotate([, b]) {
  const v = handler.bracket.number(b);
  if (v != null) {
    return [
      transformBase,
      { "--un-rotate": `${v}deg` }
    ];
  }
}

const variablesAbbrMap = {
  "visible": "visibility",
  "select": "user-select",
  "vertical": "vertical-align",
  "backface": "backface-visibility",
  "whitespace": "white-space",
  "break": "word-break",
  "color": "color",
  "case": "text-transform",
  "write": "writing-mode",
  "write-orient": "text-orientation",
  "origin": "transform-origin",
  "bg": "background-color",
  "bg-blend": "background-blend-mode",
  "bg-clip": "-webkit-background-clip",
  "bg-gradient": "linear-gradient",
  "bg-origin-border": "background-origin",
  "bg-position": "background-position",
  "bg-repeat": "background-repeat",
  "bg-size": "background-size",
  "bg-opacity": "background-opacity",
  "tab": "tab-size",
  "underline": "text-decoration-thickness",
  "underline-offset": "text-underline-offset",
  "text": "color",
  "grid-cols": "grid-template-columns",
  "grid-rows": "grid-template-rows",
  "auto-flow": "grid-auto-flow",
  "row-start": "grid-row-start",
  "row-end": "grid-row-end",
  "justify": "justify-content",
  "content": "align-content",
  "items": "align-items",
  "self": "align-self",
  "object": "object-fit",
  "mix-blend": "mix-blend-mode",
  "animate-speed": "animation-speed"
};
const cssVariables = [[
  /^(.+)-\$(.+)$/,
  ([, name, varname]) => {
    const prop = variablesAbbrMap[name];
    if (prop) {
      return {
        [prop]: `var(--${varname})`
      };
    }
  }
]];

const questionMark = [
  [
    /^(where|\?)$/,
    (_, { constructCSS, generator }) => {
      if (generator.config.envMode === "dev")
        return `@keyframes __un_qm{0%{box-shadow:inset 4px 4px #ff1e90, inset -4px -4px #ff1e90}100%{box-shadow:inset 8px 8px #3399ff, inset -8px -8px #3399ff}}
${constructCSS({ animation: "__un_qm 0.5s ease-in-out alternate infinite" })}`;
    }
  ]
];

const divideSizes = [
  [/^divide-?([xy])$/, handlerDivide],
  [/^divide-?([xy])-?(-?.+)$/, handlerDivide],
  [/^divide-?([xy])-reverse$/, ([, d]) => [[`--un-divide-${d}-reverse`, 1]]]
];
const divideStyles = [
  ["divide-solid", { "border-style": "solid" }],
  ["divide-dashed", { "border-style": "dashed" }],
  ["divide-dotted", { "border-style": "dotted" }],
  ["divide-double", { "border-style": "double" }],
  ["divide-none", { "border-style": "none" }]
];
const divides = [divideSizes, divideColors, divideStyles].flat(1);
function handlerDivide([, a, b]) {
  const [d, s = "1"] = directionMap[a] ? [a, b] : ["", a];
  const v = handler.bracket.px(s);
  if (v != null) {
    const results = directionMap[d].map((item) => {
      const key = `border${item}-width`;
      const value = item.endsWith("right") || item.endsWith("bottom") ? `calc(${v} * var(--un-divide-${d}-reverse))` : `calc(${v} * calc(1 - var(--un-divide-${d}-reverse)))`;
      return [key, value];
    });
    if (results)
      return [[`--un-divide-${d}-reverse`, 0], ...results];
  }
}

const lineClamps = [
  [/^line-clamp-(\d+)$/, ([, v]) => ({
    "overflow": "hidden",
    "display": "-webkit-box",
    "-webkit-box-orient": "vertical",
    "-webkit-line-clamp": v,
    "line-clamp": v
  })],
  ["line-clamp-none", {
    "-webkit-line-clamp": "unset",
    "line-clamp": "unset"
  }]
];

const rules = [
  screenReadersAccess,
  cssVariables,
  paddings,
  margins,
  lineClamps,
  isolations,
  container,
  displays,
  opacity,
  bgAttachments,
  bgBlendModes,
  bgClips,
  bgColors,
  bgGradients,
  bgImages,
  bgOrigins,
  bgPositions,
  bgSizes,
  bgRepeats,
  fillColors,
  borders,
  divides,
  contents,
  fonts,
  tabSizes,
  textIndents,
  textOverflows,
  textDecorations,
  textDecorationStyles,
  textDecorationColors,
  textDecorationLengths,
  textDecorationOffsets,
  textStrokeWidths,
  textStrokeColors,
  textShadows,
  textTransforms,
  textAligns,
  textColors,
  fontStyles,
  fontVariantNumeric,
  fontSmoothings,
  hyphens,
  writingModes,
  writingOrientations,
  mixBlendModes,
  boxShadows,
  rings,
  flex,
  grids,
  gaps,
  positions,
  sizes,
  aspectRatio,
  cursors,
  appearances,
  pointerEvents,
  resizes,
  verticalAligns,
  userSelects,
  whitespaces,
  listStyle,
  caretColors,
  boxDecorationBreaks,
  caretOpacity,
  imageRenderings,
  appearance,
  outline,
  placeholder,
  overscrolls,
  breaks,
  trackings,
  wordSpacings,
  leadings,
  overflows,
  positions,
  orders,
  justifies,
  justifyItems,
  justifySelfs,
  alignContents,
  alignItems,
  alignSelfs,
  placeContents,
  placeItems,
  placeSelfs,
  animations,
  insets,
  floats,
  zIndexes,
  objectPositions,
  boxSizing,
  transitions,
  filters,
  tables,
  transforms,
  questionMark
].flat(1);

export { justifySelfs as $, borders as A, parseColorUtil as B, colorResolver$2 as C, opacity as D, textColors as E, textDecorationColors as F, textStrokeColors as G, bgColors as H, borderColors as I, ringColors as J, ringOffsetColors as K, divideColors as L, fillColors as M, container as N, divideSizes as O, divideStyles as P, divides as Q, filters as R, flex as S, gaps as T, grids as U, overflows as V, lineClamps as W, positions as X, justifies as Y, orders as Z, justifyItems as _, animations as a, alignContents as a0, alignItems as a1, alignSelfs as a2, placeContents as a3, placeItems as a4, placeSelfs as a5, insets as a6, floats as a7, zIndexes as a8, objectPositions as a9, writingModes as aA, writingOrientations as aB, screenReadersAccess as aC, isolations as aD, tables as aE, transforms as aF, transitions as aG, fontsFamilies as aH, fontSizes as aI, fontWeights as aJ, leadings as aK, trackings as aL, wordSpacings as aM, tabSizes as aN, textDecorationLengths as aO, textDecorationOffsets as aP, textIndents as aQ, textStrokeWidths as aR, textShadows as aS, fontVariantNumeric as aT, fonts as aU, cssVariables as aV, boxSizing as aa, questionMark as ab, rings as ac, mixBlendModes as ad, boxShadows as ae, sizes as af, aspectRatio as ag, paddings as ah, margins as ai, varEmpty as aj, displays as ak, appearances as al, cursors as am, pointerEvents as an, resizes as ao, userSelects as ap, whitespaces as aq, contents as ar, breaks as as, textOverflows as at, textTransforms as au, textDecorations as av, textDecorationStyles as aw, fontStyles as ax, fontSmoothings as ay, hyphens as az, bgAttachments as b, containerShortcuts as c, bgBlendModes as d, bgClips as e, bgGradients as f, bgImages as g, bgOrigins as h, bgPositions as i, bgRepeats as j, bgSizes as k, listStyle as l, boxDecorationBreaks as m, caretOpacity as n, outline as o, caretColors as p, imageRenderings as q, rules as r, appearance as s, textAligns as t, placeholder as u, verticalAligns as v, overscrolls as w, borderSizes as x, borderRadius as y, borderStyles as z };
