// @unocss/preset-uno/dist/chunks/default3.mjs
import { v as variantMatcher } from './variants.mjs';
import { v as variantPseudoClasses, a as variantPseudoElements } from './pseudo.mjs';

const regexCache = {};
const variantBreakpoints = (matcher, _, theme) => {
  const variantEntries = Object.entries(theme.breakpoints || {}).map(([point, size], idx) => [point, size, idx]);
  for (const [point, size, idx] of variantEntries) {
    if (!regexCache[point])
      regexCache[point] = new RegExp(`^((?:[a|l]t-)?${point}[:-])`);
    const match = matcher.match(regexCache[point]);
    if (!match)
      continue;
    const [, pre] = match;
    let direction = "min";
    let order = parseInt(size);
    if (pre.startsWith("lt-")) {
      direction = "max";
      order = -order;
    }
    const m = matcher.slice(pre.length);
    if (m === "container")
      continue;
    if (pre.startsWith("at-") && idx < variantEntries.length - 1) {
      return {
        matcher: m,
        parent: [`@media (min-width: ${size}) and (max-width: ${variantEntries[idx + 1][1]})`, order]
      };
    }
    return {
      matcher: m,
      parent: [`@media (${direction}-width: ${size})`, order]
    };
  }
};

const variantChildren = [
  variantMatcher("children", (input) => `${input} > *`),
  variantMatcher("all", (input) => `${input} *`),
  variantMatcher("next", (input) => `${input}+*`)
];

const variantColorsClass = [
  variantMatcher("dark", (input) => `.dark $$ ${input}`),
  variantMatcher("light", (input) => `.light $$ ${input}`)
];
const variantColorsMedia = [
  (v) => {
    const dark = variantMatcher("dark")(v);
    if (dark) {
      return {
        ...dark,
        parent: "@media (prefers-color-scheme: dark)"
      };
    }
    const light = variantMatcher("light")(v);
    if (light) {
      return {
        ...light,
        parent: "@media (prefers-color-scheme: light)"
      };
    }
  }
];

const variantImportant = {
  match(matcher) {
    if (matcher.startsWith("!")) {
      return {
        matcher: matcher.slice(1),
        body: (body) => {
          body.forEach((v) => {
            if (v[1])
              v[1] += " !important";
          });
          return body;
        }
      };
    }
  }
};
const variantNegative = {
  match(matcher) {
    if (matcher.startsWith("-")) {
      return {
        matcher: matcher.slice(1),
        body: (body) => {
          body.forEach((v) => {
            if (v[0].startsWith("--un-scale") || v[1]?.toString() === "0")
              return;
            v[1] = v[1]?.toString().replace(/[0-9.]+(?:[a-z]+|%)?/, (i) => `-${i}`);
          });
          return body;
        }
      };
    }
  }
};
const variantSpace = (matcher) => {
  if (/^space-?([xy])-?(-?.+)$/.test(matcher) || /^divide-/.test(matcher)) {
    return {
      matcher,
      selector: (input) => {
        return `${input}>:not([hidden])~:not([hidden])`;
      }
    };
  }
};

const variants = [
  variantSpace,
  variantNegative,
  variantImportant,
  variantBreakpoints,
  ...variantChildren,
  variantPseudoClasses,
  variantPseudoElements
];

export { variantColorsMedia as a, variantColorsClass as b, variantBreakpoints as c, variantChildren as d, variantImportant as e, variantNegative as f, variantSpace as g, variants as v };
