// @builder/vite-service/lib/plugins/history.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serverHistoryPlugin = void 0;
var lodash_1 = require("lodash");
var deepmerge_1 = require("deepmerge");
var history = require("connect-history-api-fallback");
/**
 * Webpack devServer.historyApiFallback
 *
 * @see https://github.com/vitejs/vite/blob/8733a83d291677b9aff9d7d78797ebb44196596e/packages/vite/src/node/server/index.ts#L433
 */
var serverHistoryPlugin = function (opt) {
    if ((0, lodash_1.isBoolean)(opt))
        return null;
    return {
        name: 'vite-plugin-server-history',
        enforce: 'pre',
        configureServer: function (_a) {
            var app = _a.middlewares;
            var handler = history((0, deepmerge_1.all)([opt, {
                    rewrites: [{
                            from: /.*/,
                            to: function (ctx) { return ctx.parsedUrl.pathname; }
                        }]
                }]));
            app.use(handler);
        },
    };
};
exports.serverHistoryPlugin = serverHistoryPlugin;
