// @builder/vite-service/lib/plugins/polyfill.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.polyfillPlugin = void 0;
var fs = require("fs-extra");
var path = require("path");
var plugin_legacy_1 = require("@vitejs/plugin-legacy");
var glob = require("fast-glob");
var lodash_1 = require("lodash");
var polyfillPlugin = function (option) {
    var value = option.value, browserslist = option.browserslist, rootDir = option.rootDir, hash = option.hash, outputAssetsPath = option.outputAssetsPath;
    if (!value)
        return;
    var outDir = '';
    var assetsDirName = '';
    /**
     * hash 为开启时，清除 polyfill hash 后缀
     */
    var plugin = {
        name: 'rollup-plugin-polyfills-clear',
        config: function (cfg, _a) {
            var _b, _c, _d, _e;
            var command = _a.command;
            if (command === 'build') {
                outDir = (_c = (_b = cfg.build) === null || _b === void 0 ? void 0 : _b.outDir) !== null && _c !== void 0 ? _c : 'dist';
                assetsDirName = (_e = (_d = cfg.build) === null || _d === void 0 ? void 0 : _d.assetsDir) !== null && _e !== void 0 ? _e : 'assets';
            }
        },
        closeBundle: function () {
            return __awaiter(this, void 0, void 0, function () {
                var distPath, assetsPath, jsDirPath, polyfillsFile, hashName, fileName, dest, htmlPaths;
                return __generator(this, function (_a) {
                    distPath = path.resolve(rootDir, outDir);
                    assetsPath = path.resolve(distPath, assetsDirName);
                    jsDirPath = path.resolve(distPath, outputAssetsPath.js);
                    polyfillsFile = glob.sync(path.resolve(assetsPath, 'polyfills-legacy.*.js'))[0];
                    if (!polyfillsFile)
                        return [2 /*return*/];
                    hashName = path.relative(assetsPath, polyfillsFile);
                    fileName = hashName;
                    // 拖到 js 文件夹下
                    fs.moveSync(polyfillsFile, path.resolve(jsDirPath, fileName));
                    polyfillsFile = path.resolve(jsDirPath, fileName);
                    if (!hash) {
                        dest = path.resolve(jsDirPath, 'polyfills-legacy.js');
                        fs.renameSync(polyfillsFile, dest);
                        fileName = 'polyfills-legacy.js';
                    }
                    // remove assets dir
                    if (!fs.readdirSync(assetsPath).length) {
                        fs.rmdirSync(assetsPath);
                    }
                    htmlPaths = glob.sync(path.resolve(distPath, '*.html'));
                    htmlPaths.forEach(function (p) {
                        var html = fs.readFileSync(p, 'utf-8');
                        fs.writeFileSync(p, (0, lodash_1.replace)(html, "/" + assetsDirName + "/" + hashName, "/" + outputAssetsPath.js + "/" + fileName));
                    });
                    return [2 /*return*/];
                });
            });
        }
    };
    return [
        plugin,
        (0, plugin_legacy_1.default)({
            targets: browserslist,
            polyfills: value === 'usage'
        })
    ];
};
exports.polyfillPlugin = polyfillPlugin;
