// build-plugin-pwa/lib/ManifestPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var HtmlWebpackPlugin = require("html-webpack-plugin");
function ensureTrailingSlash(str) {
    if (str.length && str.substr(-1, 1) !== '/') {
        return str + "/";
    }
    return str;
}
var ManifestPlugin = /** @class */ (function () {
    function ManifestPlugin() {
    }
    ManifestPlugin.prototype.apply = function (compiler) {
        compiler.hooks.compilation.tap('ManifestPlugin', function (compilation) {
            var alterAssetTagsHook;
            if (HtmlWebpackPlugin.version >= 4) {
                var hooks = HtmlWebpackPlugin.getHooks(compilation);
                alterAssetTagsHook = hooks.alterAssetTags;
            }
            else {
                var hooks = compilation.hooks;
                alterAssetTagsHook = hooks.htmlWebpackPluginAlterAssetTags;
            }
            alterAssetTagsHook.tap('ManifestPlugin', function (htmlPluginData) {
                var _a;
                var publicPath = ((_a = compilation === null || compilation === void 0 ? void 0 : compilation.outputOptions) === null || _a === void 0 ? void 0 : _a.publicPath) ? compilation.outputOptions.publicPath : '/';
                // append publicPath to htm
                if (htmlPluginData.assetTags) {
                    htmlPluginData.assetTags.meta.push({
                        tagName: 'link',
                        voidTag: true,
                        attributes: { rel: 'manifest', href: ensureTrailingSlash(publicPath) + "manifest.json" }
                    });
                }
            });
        });
    };
    return ManifestPlugin;
}());
exports.default = ManifestPlugin;
