// build-scripts/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.jestTest = exports.webpackBuild = exports.webpackStart = exports.Context = exports.JestService = exports.WebpackService = exports.test = exports.start = exports.build = void 0;
const build = require("./commands/build");
exports.build = build;
const start = require("./commands/start");
exports.start = start;
const test = require("./commands/test");
exports.test = test;
const webpackStart = require("./service/start");
exports.webpackStart = webpackStart;
const webpackBuild = require("./service/build");
exports.webpackBuild = webpackBuild;
const jestTest = require("./service/test");
exports.jestTest = jestTest;
const WebpackService_1 = require("./service/WebpackService");
exports.WebpackService = WebpackService_1.default;
const JestService_1 = require("./service/JestService");
exports.JestService = JestService_1.default;
const Context_1 = require("./core/Context");
exports.Context = Context_1.default;
__exportStar(require("./core/Context"), exports);
__exportStar(require("./types"), exports);
