// build-plugin-ice-logger/lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var levels_1 = require("./levels");
var plugin = function (_a) {
    var applyMethod = _a.applyMethod, registerUserConfig = _a.registerUserConfig, onGetWebpackConfig = _a.onGetWebpackConfig, context = _a.context;
    return __awaiter(void 0, void 0, void 0, function () {
        var exportName;
        return __generator(this, function (_b) {
            registerUserConfig([
                {
                    name: 'dropLogLevel',
                    validation: 'string'
                }
            ]);
            exportName = 'logger';
            applyMethod('addPluginTemplate', path.join(__dirname, "../" + exportName));
            // add ice exports
            applyMethod('addExport', {
                source: "./plugins/" + exportName,
                exportName: exportName,
                importSource: "$$ice/plugins/" + exportName,
                exportDefault: exportName,
            });
            onGetWebpackConfig(function (config) {
                var dropLogLevel = context.userConfig.dropLogLevel;
                if (dropLogLevel) {
                    var pureFuncs_1 = getPureFuncs(levels_1.default[dropLogLevel.toLowerCase()]);
                    // @ts-ignore
                    if (config.optimization.minimizers.has('ESBuild')) {
                        config.optimization.minimizer('ESBuild').tap(function (_a) {
                            var options = _a[0];
                            return [
                                __assign(__assign({}, options), { pure: pureFuncs_1 })
                            ];
                        });
                        // @ts-ignore
                    }
                    else if (config.optimization.minimizers.has('TerserPlugin')) {
                        config.optimization.minimizer('TerserPlugin').tap(function (_a) {
                            var options = _a[0];
                            return [
                                __assign(__assign({}, options), { terserOptions: __assign(__assign({}, options.terserOptions), { compress: __assign(__assign({}, options.terserOptions.compress), { drop_console: false, pure_funcs: pureFuncs_1 }) }) })
                            ];
                        });
                        // @ts-ignore
                    }
                    else if (config.optimization.minimizers.has('SWC')) {
                        config.optimization.minimizer('SWC').tap(function (_a) {
                            var _b, _c, _d;
                            var options = _a[0];
                            return [
                                __assign(__assign({}, options), { jsc: __assign(__assign({}, options.jsc), { minify: __assign(__assign({}, (_b = options.jsc) === null || _b === void 0 ? void 0 : _b.minify), { compress: __assign(__assign({}, (_d = (_c = options.jsc) === null || _c === void 0 ? void 0 : _c.minify) === null || _d === void 0 ? void 0 : _d.compress), { drop_console: false, pure_funcs: pureFuncs_1 }) }) }) })
                            ];
                        });
                    }
                }
            });
            return [2 /*return*/];
        });
    });
};
function getPureFuncs(level) {
    return Object.keys(levels_1.default)
        .filter(function (methodName) { return levels_1.default[methodName] <= level; })
        .map(function (methodName) { return "console." + methodName; });
}
exports.default = plugin;
