// build-plugin-app-core/lib/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExportApiKeys = exports.EXPORT_API_MPA = exports.ICE_TEMP = exports.TEMP_PATH = exports.PROJECT_TYPE = exports.USER_CONFIG = void 0;
var validation = require('@builder/app-helpers').validation;
exports.USER_CONFIG = [
    {
        name: 'store',
        validation: function (val) {
            return validation('store', val, 'object|boolean');
        }
    },
    {
        name: 'ssr',
        validation: 'boolean|string'
    },
    {
        name: 'auth',
        validation: 'boolean'
    },
    {
        name: 'sourceDir',
        validation: 'string',
    },
    {
        // add generateRuntime in case of runtimes do not pass the ts checker
        name: 'generateRuntime',
        validation: 'boolean',
        defaultValue: false,
    }
];
exports.PROJECT_TYPE = 'PROJECT_TYPE';
exports.TEMP_PATH = 'TEMP_PATH';
exports.ICE_TEMP = 'ICE_TEMP';
/**
 * API_Names
 *
 * deprecated api
 * addIceExport、addIceTypesExport、addIceAppConfigTypes、addIceAppConfigAppTypes
 *
 * for ice and rax
 * addExport、addTypesExport、addAppConfigTypes、addAppConfigAppTypes
 */
exports.EXPORT_API_MPA = [
    {
        name: ['addIceExport', 'addExport'],
        value: ['imports', 'exports']
    },
    {
        name: ['addIceTypesExport', 'addTypesExport'],
        value: ['typesImports', 'typesExports'],
    },
    {
        name: ['addIceAppConfigTypes', 'addAppConfigTypes'],
        value: ['appConfigTypesImports', 'appConfigTypesExports']
    },
    {
        name: ['addIceAppConfigAppTypes', 'addAppConfigAppTypes'],
        value: ['appConfigAppTypesImports', 'appConfigAppTypesExports']
    }
];
var getExportApiKeys = function () {
    var apiKeys = [];
    exports.EXPORT_API_MPA.forEach(function (item) {
        apiKeys = apiKeys.concat(item.name);
    });
    return apiKeys;
};
exports.getExportApiKeys = getExportApiKeys;
