// @builder/user-config/lib/Loaders/PreCompileLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/no-dynamic-require */
var swc_1 = require("@builder/swc");
var loader_utils_1 = require("loader-utils");
var deepmerge = require("deepmerge");
function default_1(source) {
    var callback = this.async();
    var filename = this.resourcePath;
    var devtool = this._compiler.options.devtool;
    // Define sourceMaps by webpack devtool values
    var loaderOptions = (0, loader_utils_1.getOptions)(this) || {};
    var sourceMaps = !!devtool;
    var compileOptions = deepmerge({
        filename: filename,
        sourceFileName: filename,
        sourceMaps: sourceMaps,
        jsc: {
            parser: {
                syntax: 'typescript',
                tsx: /\.tsx/.test(filename),
            },
            target: 'es2021'
        },
    }, loaderOptions);
    try {
        var _a = (0, swc_1.transformSync)(source, compileOptions), code = _a.code, map = _a.map;
        callback(null, code, map);
    }
    catch (err) {
        callback(err);
    }
}
exports.default = default_1;
