// @builder/user-config/lib/Loaders/RedirectPathLoader.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
var app_helpers_1 = require("@builder/app-helpers");
var loader_utils_1 = require("loader-utils");
function default_1(code, sourcemap) {
    // Make the loader async
    var callback = this.async();
    this.cacheable();
    loader.call(this, code, sourcemap).then(function (args) { return callback.apply(void 0, __spreadArray([null], args, false)); }, function (err) { return callback(err); });
}
exports.default = default_1;
function loader(code, sourcemap) {
    return __awaiter(this, void 0, void 0, function () {
        var options, sources, _i, sources_1, source, redirects;
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    options = (0, loader_utils_1.getOptions)(this);
                    sources = Object.keys(options);
                    _i = 0, sources_1 = sources;
                    _a.label = 1;
                case 1:
                    if (!(_i < sources_1.length)) return [3 /*break*/, 4];
                    source = sources_1[_i];
                    redirects = options[source];
                    if (!redirects) return [3 /*break*/, 3];
                    return [4 /*yield*/, (0, app_helpers_1.redirectImport)(code, {
                            source: source,
                            redirectImports: redirects
                                .map(function (redirect) {
                                var name = redirect.name, value = redirect.value, type = redirect.type, alias = redirect.alias, _a = redirect.multipleSource, multipleSource = _a === void 0 ? [] : _a;
                                var matchedSource = multipleSource.find(function (_a) {
                                    var filename = _a.filename;
                                    return filename === _this.resourcePath;
                                });
                                if (matchedSource) {
                                    return {
                                        name: name,
                                        redirectPath: matchedSource.value,
                                        default: matchedSource.type === 'default',
                                    };
                                }
                                return {
                                    name: name,
                                    redirectPath: alias || value,
                                    default: type === 'default',
                                };
                            }),
                        })];
                case 2:
                    // eslint-disable-next-line no-await-in-loop
                    code = _a.sent();
                    _a.label = 3;
                case 3:
                    _i++;
                    return [3 /*break*/, 1];
                case 4: return [2 /*return*/, [code, sourcemap]];
            }
        });
    });
}
