// build-plugin-app-core/lib/config/setRegisterMethod.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var getPages_1 = require("../utils/getPages");
var getRoutes_1 = require("../utils/getRoutes");
var formatPath_1 = require("../utils/formatPath");
var getSourceDir_1 = require("../utils/getSourceDir");
var formatPluginDir_1 = require("../utils/formatPluginDir");
var getBuildConfig_1 = require("../utils/getBuildConfig");
var getSourceFile_1 = require("../utils/getSourceFile");
var constant_1 = require("../constant");
var importDeclarations_1 = require("./importDeclarations");
exports.default = (function (api, options) {
    var registerMethod = api.registerMethod;
    var generator = options.generator;
    // register utils method
    registerMethod('getPages', getPages_1.default);
    registerMethod('formatPath', formatPath_1.default);
    registerMethod('getRoutes', getRoutes_1.default);
    registerMethod('getSourceDir', getSourceDir_1.default);
    registerMethod('getBuildConfig', getBuildConfig_1.default);
    registerMethod('getSourceFile', getSourceFile_1.default);
    // registerMethod for modify page
    registerMethod('addPageExport', generator.addPageExport);
    registerMethod('removePageExport', generator.removePageExport);
    // registerMethod for render content
    registerMethod('addRenderFile', generator.addRenderFile);
    registerMethod('addPluginTemplate', function (pluginName) { return function (templateOptions, extraData) {
        var _a = typeof templateOptions === 'string'
            ? { template: templateOptions, targetDir: (0, formatPluginDir_1.default)(pluginName) }
            : templateOptions, template = _a.template, targetDir = _a.targetDir;
        generator.addTemplateFiles({
            template: template,
            // overwrite target dir, build-plugin-ice-request => request
            targetDir: "plugins/" + targetDir,
        }, extraData);
    }; }, { pluginName: true });
    registerMethod('modifyRenderData', generator.modifyRenderData.bind(generator));
    registerMethod('addDisableRuntimePlugin', generator.addDisableRuntimePlugin);
    function addImportDeclaration(data) {
        var importSource = data.importSource, exportMembers = data.exportMembers, exportDefault = data.exportDefault, alias = data.alias, multipleSource = data.multipleSource;
        if (importSource) {
            if (exportMembers) {
                exportMembers.forEach(function (exportMember) {
                    // import { withAuth } from 'ice' -> import { withAuth } from 'ice/auth';
                    importDeclarations_1.default[exportMember] = __assign(__assign({}, importDeclarations_1.default[exportMember]), { value: importSource, type: 'normal' });
                });
            }
            else if (exportDefault) {
                // import { Helmet } from 'ice' -> import Helmet from 'ice/helmet';
                importDeclarations_1.default[exportDefault] = __assign(__assign({}, importDeclarations_1.default[exportDefault]), { value: importSource, type: 'default' });
            }
            if (alias) {
                Object.keys(alias).forEach(function (exportMember) {
                    // import { Head } from 'ice'; -> import { Helmet as Head } from 'react-helmet';
                    importDeclarations_1.default[exportMember] = __assign(__assign({}, importDeclarations_1.default[exportMember]), { value: importSource, type: 'normal', alias: alias[exportMember] });
                });
            }
        }
        if (multipleSource) {
            Object.keys(multipleSource).forEach(function (exportMember) {
                importDeclarations_1.default[exportMember] = __assign(__assign({}, importDeclarations_1.default[exportMember]), { multipleSource: multipleSource[exportMember] });
            });
        }
    }
    registerMethod('addImportDeclaration', addImportDeclaration);
    api.setValue('importDeclarations', importDeclarations_1.default);
    // registerMethod for add export
    var apiKeys = (0, constant_1.getExportApiKeys)();
    apiKeys.forEach(function (apiKey) {
        registerMethod(apiKey, function (exportData) {
            addImportDeclaration(exportData);
            generator.addExport(apiKey, exportData);
        });
        registerMethod(apiKey.replace('add', 'remove'), function (removeExportName) {
            generator.removeExport(apiKey, removeExportName);
        });
    });
    var registerAPIs = {
        addEntryImports: {
            apiKey: 'addContent',
        },
        addEntryCode: {
            apiKey: 'addContent',
        },
    };
    Object.keys(registerAPIs).forEach(function (apiName) {
        registerMethod(apiName, function (code, position) {
            if (position === void 0) { position = 'after'; }
            var apiKey = registerAPIs[apiName].apiKey;
            generator[apiKey](apiName, code, position);
        });
    });
});
